<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Sofia Salazar">
<meta name="dcterms.date" content="2023-09-24">
<meta name="description" content="Using Borzoi for predicting on a personal genome, derived from VCF">

<title>Borzoi prediction from personal genome - jupyter notebook – GENE 46100</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-30950e57ca2d683c179b59b95dd53cbb.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">GENE 46100</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#set-up" id="toc-set-up" class="nav-link active" data-scroll-target="#set-up">1. Set up</a>
  <ul class="collapse">
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load libraries</a></li>
  <li><a href="#configure-gpu-device" id="toc-configure-gpu-device" class="nav-link" data-scroll-target="#configure-gpu-device">Configure GPU device</a></li>
  <li><a href="#dowload-model-files" id="toc-dowload-model-files" class="nav-link" data-scroll-target="#dowload-model-files">Dowload model files</a></li>
  <li><a href="#download-reference-genome" id="toc-download-reference-genome" class="nav-link" data-scroll-target="#download-reference-genome">Download reference genome</a></li>
  <li><a href="#define-paths" id="toc-define-paths" class="nav-link" data-scroll-target="#define-paths">Define paths</a></li>
  </ul></li>
  <li><a href="#define-functions" id="toc-define-functions" class="nav-link" data-scroll-target="#define-functions">2. Define functions</a>
  <ul class="collapse">
  <li><a href="#extract-fasta" id="toc-extract-fasta" class="nav-link" data-scroll-target="#extract-fasta">Extract fasta</a></li>
  </ul></li>
  <li><a href="#model-configuration" id="toc-model-configuration" class="nav-link" data-scroll-target="#model-configuration">3.Model configuration</a></li>
  <li><a href="#build-sequence" id="toc-build-sequence" class="nav-link" data-scroll-target="#build-sequence">4. Build sequence</a></li>
  <li><a href="#predict" id="toc-predict" class="nav-link" data-scroll-target="#predict">5. Predict</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">Borzoi prediction from personal genome - jupyter notebook</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li></ul></div></div>
  <div class="quarto-categories">
    <div class="quarto-category">gene46100</div>
    <div class="quarto-category">notebook</div>
  </div>
  </div>

<div>
  <div class="description">
    Using Borzoi for predicting on a personal genome, derived from VCF
  </div>
</div>


<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Sofia Salazar </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 24, 2023</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="set-up" class="level2">
<h2 class="anchored" data-anchor-id="set-up">1. Set up</h2>
<p>Recommended: create a new conda environment, borzoi is not compatible with python &gt; 3.10</p>
<p>conda create –name borzoi46100 python=3.10</p>
<p>Then, use borzoi46100 environment to run the notebook</p>
<p>Install software repositories</p>
<div id="cell-4" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="9e3192ea-14fe-414f-9b7b-881f580f6fa2" data-execution_count="1">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>GITHUB_DIR <span class="op">=</span> <span class="st">'/Users/haekyungim/Github'</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># clone baskerville and borzoi if not already cloned</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>baskerville_path <span class="op">=</span> os.path.join(GITHUB_DIR, <span class="st">'baskerville'</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>borzoi_path <span class="op">=</span> os.path.join(GITHUB_DIR, <span class="st">'borzoi'</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(baskerville_path):</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>git clone https:<span class="op">//</span>github.com<span class="op">/</span>calico<span class="op">/</span>baskerville.git {GITHUB_DIR}<span class="op">/</span>baskerville</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(borzoi_path):</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>git clone https:<span class="op">//</span>github.com<span class="op">/</span>calico<span class="op">/</span>borzoi.git {GITHUB_DIR}<span class="op">/</span>borzoi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><strong>After loading baskerville, restart runtime, run code from here</strong></p>
<p>Install libraries</p>
<div id="cell-7" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="56e21a62-7fec-43fa-cac9-f30c402cdaa0" data-execution_count="2">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> baskerville</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">ImportError</span>:</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>pip install <span class="op">-</span>e {GITHUB_DIR}<span class="op">/</span>baskerville</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-8" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="811f86be-4d60-43c9-d103-f2177c7a6356" data-execution_count="3">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> borzoi</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">ImportError</span>:</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>pip install <span class="op">-</span>e {GITHUB_DIR}<span class="op">/</span>borzoi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>NOTE: install tensorflow-metal if running on a mac</p>
<div id="cell-10" class="cell" data-execution_count="7">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import platform</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># if platform.processor() == 'arm':</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a><span class="co">#     print("Apple Silicon detected, installing tensorflow-metal...")</span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co">#     %pip install tensorflow-metal</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># else:</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a><span class="co">#     print("Not running on Apple Silicon, skipping tensorflow-metal installation")</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<div class="ansi-escaped-output">
<pre>Apple Silicon detected, installing tensorflow-metal...

Collecting tensorflow-metal

  Downloading tensorflow_metal-1.2.0-cp310-cp310-macosx_12_0_arm64.whl.metadata (1.3 kB)

Requirement already satisfied: wheel~=0.35 in /Users/haekyungim/miniconda3/envs/borzoi46100/lib/python3.10/site-packages (from tensorflow-metal) (0.45.1)

Requirement already satisfied: six&gt;=1.15.0 in /Users/haekyungim/miniconda3/envs/borzoi46100/lib/python3.10/site-packages (from tensorflow-metal) (1.17.0)

Downloading tensorflow_metal-1.2.0-cp310-cp310-macosx_12_0_arm64.whl (1.4 MB)

   <span class="ansi-bright-black-fg">━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━</span> <span class="ansi-green-fg">1.4/1.4 MB</span> <span class="ansi-red-fg">23.8 MB/s</span> eta <span class="ansi-cyan-fg">0:00:00</span>

Installing collected packages: tensorflow-metal

Successfully installed tensorflow-metal-1.2.0

Note: you may need to restart the kernel to use updated packages.
</pre>
</div>
</div>
</div>
<section id="load-libraries" class="level3">
<h3 class="anchored" data-anchor-id="load-libraries">Load libraries</h3>
<div id="cell-12" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="125ea458-d0df-4254-8111-e55225a2c0d3" data-execution_count="4">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> kipoiseq</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">ImportError</span>:</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">%</span>pip install kipoiseq</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> kipoiseq <span class="im">import</span> Interval</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a><span class="cf">try</span>:</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> cyvcf2</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="cf">except</span> <span class="pp">ImportError</span>:</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    <span class="op">%</span>pip install cyvcf2</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> time</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> io</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> gzip</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a><span class="co">#os.environ['CUDA_VISIBLE_DEVICES'] = '-1'</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> h5py</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> tensorflow <span class="im">as</span> tf</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> baskerville</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> baskerville <span class="im">import</span> seqnn</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> baskerville <span class="im">import</span> gene <span class="im">as</span> bgene</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> baskerville <span class="im">import</span> dna</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pysam</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pyfaidx</span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.patches <span class="im">as</span> patches</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="configure-gpu-device" class="level3">
<h3 class="anchored" data-anchor-id="configure-gpu-device">Configure GPU device</h3>
<div id="cell-14" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="2ad0c701-ebb7-42a9-8f0a-a6787699465f" data-execution_count="5">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"Num GPUs Available: "</span>, <span class="bu">len</span>(tf.config.list_physical_devices(<span class="st">'GPU'</span>)))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>gpu_device <span class="op">=</span> tf.config.list_physical_devices(<span class="st">'GPU'</span>)[<span class="dv">0</span>]</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>tf.config.experimental.set_memory_growth(gpu_device, <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Num GPUs Available:  1</code></pre>
</div>
</div>
</section>
<section id="dowload-model-files" class="level3">
<h3 class="anchored" data-anchor-id="dowload-model-files">Dowload model files</h3>
<div id="cell-16" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="ab55c597-0013-4644-be54-8970cfeb6bbe" data-execution_count="8">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>PRE <span class="op">=</span> <span class="st">'/Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100'</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>CONTENT_DIR <span class="op">=</span> PRE <span class="op">+</span> <span class="st">'/borzoi'</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create borzoi/data directory if it doesn't exist</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(CONTENT_DIR)):</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {CONTENT_DIR}</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(CONTENT_DIR, <span class="st">'/data'</span>)):</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {CONTENT_DIR}<span class="op">/</span>data</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Create model file structure</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>saved_models_path <span class="op">=</span> os.path.join(CONTENT_DIR, <span class="st">'data/saved_models'</span>)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(saved_models_path):</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {saved_models_path}</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f0'</span>)):</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {saved_models_path}<span class="op">/</span>f0</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f1'</span>)):</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {saved_models_path}<span class="op">/</span>f1</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f2'</span>)):</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {saved_models_path}<span class="op">/</span>f2</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f3'</span>)):</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>mkdir {saved_models_path}<span class="op">/</span>f3</span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a><span class="co">#%cd {CONTENT_DIR}/data</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>mkdir: /Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100/borzoi/data: File exists
/Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100/borzoi/data</code></pre>
</div>
</div>
<div id="cell-17" class="cell" data-execution_count="14">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co">#DELET</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>os.path.join(CONTENT_DIR, <span class="st">'data/hg38.fa'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>'/Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100/borzoi/data/hg38.fa'</code></pre>
</div>
</div>
<div id="cell-18" class="cell" data-execution_count="11">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Download model weights</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f0'</span>, <span class="st">'model0_best.h5'</span>)):</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>borzoi<span class="op">/</span>f0<span class="op">/</span>model0_best.h5 <span class="op">-</span>O {saved_models_path}<span class="op">/</span>f0<span class="op">/</span>model0_best.h5</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f1'</span>, <span class="st">'model0_best.h5'</span>)): </span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>borzoi<span class="op">/</span>f1<span class="op">/</span>model0_best.h5 <span class="op">-</span>O {saved_models_path}<span class="op">/</span>f1<span class="op">/</span>model0_best.h5</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f2'</span>, <span class="st">'model0_best.h5'</span>)):</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>borzoi<span class="op">/</span>f2<span class="op">/</span>model0_best.h5 <span class="op">-</span>O {saved_models_path}<span class="op">/</span>f2<span class="op">/</span>model0_best.h5</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(saved_models_path, <span class="st">'f3'</span>, <span class="st">'model0_best.h5'</span>)):</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>borzoi<span class="op">/</span>f3<span class="op">/</span>model0_best.h5 <span class="op">-</span>O {saved_models_path}<span class="op">/</span>f3<span class="op">/</span>model0_best.h5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-19" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="ed59c16f-294c-4392-d53a-30bb2c33be53" data-execution_count="28">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Download and uncompress annotation files</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(CONTENT_DIR, <span class="st">'data'</span>, <span class="st">'gencode41_basic_nort.gtf'</span>)):</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget  <span class="op">-</span>O <span class="op">-</span> https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>helper<span class="op">/</span>gencode41_basic_nort.gtf.gz <span class="op">|</span> gunzip <span class="op">-</span>c <span class="op">&gt;</span> {os.path.join(CONTENT_DIR, <span class="st">'data'</span>, <span class="st">'gencode41_basic_nort.gtf'</span>)}</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(CONTENT_DIR, <span class="st">'data'</span>, <span class="st">'gencode41_basic_protein_splice.csv.gz'</span>)):</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>helper<span class="op">/</span>gencode41_basic_protein_splice.csv.gz <span class="op">-</span>O {os.path.join(CONTENT_DIR, <span class="st">'data'</span>, <span class="st">'gencode41_basic_protein_splice.csv.gz'</span>)}</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(os.path.join(CONTENT_DIR, <span class="st">'data'</span>, <span class="st">'polyadb_human_v3.csv.gz'</span>)):</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>storage.googleapis.com<span class="op">/</span>seqnn<span class="op">-</span>share<span class="op">/</span>helper<span class="op">/</span>polyadb_human_v3.csv.gz <span class="op">-</span>O {os. path.join(CONTENT_DIR, <span class="st">'data'</span>, <span class="st">'polyadb_human_v3.csv.gz'</span>)}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="download-reference-genome" class="level3">
<h3 class="anchored" data-anchor-id="download-reference-genome">Download reference genome</h3>
<div id="cell-21" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="5e2ff7a0-3391-4bfb-94c4-384b0f064329" data-execution_count="15">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co">#fasta_file = '/content/data/genome.fa'</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>fasta_file <span class="op">=</span> os.path.join(CONTENT_DIR, <span class="st">'data/hg38.fa'</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(fasta_file):</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget <span class="op">-</span>O <span class="op">-</span> http:<span class="op">//</span>hgdownload.cse.ucsc.edu<span class="op">/</span>goldenPath<span class="op">/</span>hg38<span class="op">/</span>bigZips<span class="op">/</span>hg38.fa.gz <span class="op">|</span> gunzip <span class="op">-</span>c <span class="op">&gt;</span> {fasta_file}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-22" class="cell" data-execution_count="17">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co">### download vcf file and index for chromosome 22</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>VCF_FILE <span class="op">=</span> os.path.join(CONTENT_DIR, <span class="st">'data/ALL.chr22.shapeit2_integrated_SNPs_v2a_27022019.GRCh38.phased.gz'</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> <span class="kw">not</span> os.path.exists(VCF_FILE):</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>    <span class="op">!</span>wget https:<span class="op">//</span>uchicago.box.com<span class="op">/</span>shared<span class="op">/</span>static<span class="op">/</span>u526pjh29w93ufn65yomlch4z8lkb1sp.gz <span class="op">-</span>O {VCF_FILE}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>--2025-05-04 23:18:01--  https://uchicago.box.com/shared/static/u526pjh29w93ufn65yomlch4z8lkb1sp.gz
Resolving uchicago.box.com (uchicago.box.com)... 74.112.186.157
Connecting to uchicago.box.com (uchicago.box.com)|74.112.186.157|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /public/static/u526pjh29w93ufn65yomlch4z8lkb1sp.gz [following]
--2025-05-04 23:18:05--  https://uchicago.box.com/public/static/u526pjh29w93ufn65yomlch4z8lkb1sp.gz
Reusing existing connection to uchicago.box.com:443.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://uchicago.app.box.com/public/static/u526pjh29w93ufn65yomlch4z8lkb1sp.gz [following]
--2025-05-04 23:18:06--  https://uchicago.app.box.com/public/static/u526pjh29w93ufn65yomlch4z8lkb1sp.gz
Resolving uchicago.app.box.com (uchicago.app.box.com)... 74.112.186.157
Connecting to uchicago.app.box.com (uchicago.app.box.com)|74.112.186.157|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://public.boxcloud.com/d/1/b1!RHvdIthE36FYyn0QLtnT0Bd4TIsiBQIoBEWvZGN7fVlSkjJkj5zRGnxAgcyWT75KdCvMqjeykC5OtsuTJP9LtQazK-z4ZyX6W9K4IWYfpcPQ67INeF0g8ST99TB4sR-fsAZxsGiDA80d3RwZom5lXnpC8__U5y9y5MkwgaaUV14pLDVPP7tNkFtPppWdkrcIo_78YFKZ5hB2J4ZqdH1_vZsfAVD7HeXu69e8rmSIeiSrGI2RlvPRH6ArsbuJSmMQ6GTjHLCgeAW2zq-EwzqvWYVvyra6-BfS1WGFZ02PiLIsSP4mxL9l7BR95kJqzk9zTmejGYBVxFGX1Q2fjpbQtKBBRxx0wjsnUwlQ1dzV3I251e5PSRgsZt-Ln03YYtOQZCiCLa7tPHt1Lj-5Zq_RS-8uPqaEaIaVxO3BL9H1XLOCvo16rQF5TWJ5cP5uv49l6z1llm37VYm-E_9wKg3r-VHqGZqJDcoB-EObe-UqMgSn0dQJyzsMmeI4eYhx8crMaGYFP18HGeA403GYC_VAqZ3T7dJnxyxAOCnL_3tvzJO9n5zNb6cDCuUSO6jvCKYuQ62fDgqKdtGK4Gc3BTfLZt6ikXKBTdp3Iqv9fxV8c3ecSV92UWyM95YiFftzYDzYKK8K4Qg68JhPE8BGDQCaElazdGWUHSv6aG9FofMB-YInY2YexGCeYWA5VRFkaUSEFv19TtqLjUB_zksXQQdfhnm9D9Ri2aV1iZSw7EhcI8SX7BbHxWp5-4KaVdQ4Da2_iq-mp4dzQUOCiZBULNIcAov1Upu3pSOGmSqTDuuAGm2a9yakG-D__hiSoPL2W-3gKX2XOEqqAypS10cVq0iE7WfDro0xf3el1UfN9DnSBx_i1lkspAF1gwWoMBbt-PCjTQn1QrC4brBNsHXXgs-4YPiJhtxitocc3rmQ7QiBpwufdihSnEb-gxxQwP_P2ixrVpcBn4GEfErjkpOYrjN2Kvuedu_raeWS2c8Ek2sQkJpM9A0aiwxblg-ZRjFszKntmL9auguHq5l8QKVCcvz5Okr49dpd1dxTZsK4HCC0DuclTebjWKUmE3d5jZDk5zM6aJ2NcADklr5H-cuZSi4fdbcCHeufIE92k_6bahN9WK0MQrcv4yTlKsGkaveWgxDpu4tNqnFTe_ckgr6OWzppsITVOiBMO5SilHeCRNfaBdSXIPZ4R76ATrIySz51YLGCsQpkeB-qXEgAMXDIQaKHkN0_wpqsQlJlsXOchvJ3Zs_g8ZtooEM14-sYhP7jwDYK30_Gq5PMAhYBTT33XUh_NOJiHOLFU-3jgi7thJ_AeMBWfgedpXC0coX7isbLsqhPOXewyWmQxKFxauYz7jv6GmomfUgsmAcAuMKmKcbUZX10DfWgvlH3LYGgfdLsXuDY8j3B2On0Ya_mMfkLZVgR8enXtrvQzCjV1SKsh85oVPUJF_GVojV_MtEgIHINRfK_DuqhO6LfMLjSzwCKfdN0C8vBYluHH_bIgTvVm9SkasNQCb0hpnRdrgpNXDp0M49qHok1UCFmbT1xHWUk7_ZSWPerEpUQQ9dRM8ngDU6pq6FgLaKiM8ksIIF4VZMg/download [following]
--2025-05-04 23:18:10--  https://public.boxcloud.com/d/1/b1!RHvdIthE36FYyn0QLtnT0Bd4TIsiBQIoBEWvZGN7fVlSkjJkj5zRGnxAgcyWT75KdCvMqjeykC5OtsuTJP9LtQazK-z4ZyX6W9K4IWYfpcPQ67INeF0g8ST99TB4sR-fsAZxsGiDA80d3RwZom5lXnpC8__U5y9y5MkwgaaUV14pLDVPP7tNkFtPppWdkrcIo_78YFKZ5hB2J4ZqdH1_vZsfAVD7HeXu69e8rmSIeiSrGI2RlvPRH6ArsbuJSmMQ6GTjHLCgeAW2zq-EwzqvWYVvyra6-BfS1WGFZ02PiLIsSP4mxL9l7BR95kJqzk9zTmejGYBVxFGX1Q2fjpbQtKBBRxx0wjsnUwlQ1dzV3I251e5PSRgsZt-Ln03YYtOQZCiCLa7tPHt1Lj-5Zq_RS-8uPqaEaIaVxO3BL9H1XLOCvo16rQF5TWJ5cP5uv49l6z1llm37VYm-E_9wKg3r-VHqGZqJDcoB-EObe-UqMgSn0dQJyzsMmeI4eYhx8crMaGYFP18HGeA403GYC_VAqZ3T7dJnxyxAOCnL_3tvzJO9n5zNb6cDCuUSO6jvCKYuQ62fDgqKdtGK4Gc3BTfLZt6ikXKBTdp3Iqv9fxV8c3ecSV92UWyM95YiFftzYDzYKK8K4Qg68JhPE8BGDQCaElazdGWUHSv6aG9FofMB-YInY2YexGCeYWA5VRFkaUSEFv19TtqLjUB_zksXQQdfhnm9D9Ri2aV1iZSw7EhcI8SX7BbHxWp5-4KaVdQ4Da2_iq-mp4dzQUOCiZBULNIcAov1Upu3pSOGmSqTDuuAGm2a9yakG-D__hiSoPL2W-3gKX2XOEqqAypS10cVq0iE7WfDro0xf3el1UfN9DnSBx_i1lkspAF1gwWoMBbt-PCjTQn1QrC4brBNsHXXgs-4YPiJhtxitocc3rmQ7QiBpwufdihSnEb-gxxQwP_P2ixrVpcBn4GEfErjkpOYrjN2Kvuedu_raeWS2c8Ek2sQkJpM9A0aiwxblg-ZRjFszKntmL9auguHq5l8QKVCcvz5Okr49dpd1dxTZsK4HCC0DuclTebjWKUmE3d5jZDk5zM6aJ2NcADklr5H-cuZSi4fdbcCHeufIE92k_6bahN9WK0MQrcv4yTlKsGkaveWgxDpu4tNqnFTe_ckgr6OWzppsITVOiBMO5SilHeCRNfaBdSXIPZ4R76ATrIySz51YLGCsQpkeB-qXEgAMXDIQaKHkN0_wpqsQlJlsXOchvJ3Zs_g8ZtooEM14-sYhP7jwDYK30_Gq5PMAhYBTT33XUh_NOJiHOLFU-3jgi7thJ_AeMBWfgedpXC0coX7isbLsqhPOXewyWmQxKFxauYz7jv6GmomfUgsmAcAuMKmKcbUZX10DfWgvlH3LYGgfdLsXuDY8j3B2On0Ya_mMfkLZVgR8enXtrvQzCjV1SKsh85oVPUJF_GVojV_MtEgIHINRfK_DuqhO6LfMLjSzwCKfdN0C8vBYluHH_bIgTvVm9SkasNQCb0hpnRdrgpNXDp0M49qHok1UCFmbT1xHWUk7_ZSWPerEpUQQ9dRM8ngDU6pq6FgLaKiM8ksIIF4VZMg/download
Resolving public.boxcloud.com (public.boxcloud.com)... 74.112.186.164
Connecting to public.boxcloud.com (public.boxcloud.com)|74.112.186.164|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 374610665 (357M) [application/octet-stream]
Saving to: ‘/Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100/borzoi/data/ALL.chr22.shapeit2_integrated_SNPs_v2a_27022019.GRCh38.phased.gz’

/Users/haekyungim/L 100%[===================&gt;] 357.26M  97.6MB/s    in 4.0s    

2025-05-04 23:18:19 (90.1 MB/s) - ‘/Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100/borzoi/data/ALL.chr22.shapeit2_integrated_SNPs_v2a_27022019.GRCh38.phased.gz’ saved [374610665/374610665]
</code></pre>
</div>
</div>
<div id="cell-23" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="5ebabc68-7d50-420c-d7c0-169f64b20e34" data-execution_count="18">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">#!ls /content/data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>ls {CONTENT_DIR}<span class="op">/</span>data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<div class="ansi-escaped-output">
<pre>ALL.chr22.shapeit2_integrated_SNPs_v2a_27022019.GRCh38.phased.gz

gencode41_basic_nort.gtf

gencode41_basic_nort.gtf.gz

gencode41_basic_protein_splice.csv.gz

hg38.fa

polyadb_human_v3.csv.gz

<span class="ansi-blue-fg">saved_models</span>
</pre>
</div>
</div>
</div>
<div id="cell-24" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="0a10b962-384f-4cc4-ffae-9bd985b438df" data-execution_count="19">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>cd {CONTENT_DIR}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>/Users/haekyungim/Library/CloudStorage/Box-Box/LargeFiles/imlab-data/data-Github/web-data/web-GENE-46100/borzoi</code></pre>
</div>
</div>
<section id="download-htslib-to-index-genome" class="level4">
<h4 class="anchored" data-anchor-id="download-htslib-to-index-genome">Download htslib to index genome</h4>
<div id="cell-26" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="a2f94259-2b87-4b17-fd2d-f1c2ce3a41c4">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>wget https:<span class="op">//</span>github.com<span class="op">/</span>samtools<span class="op">/</span>htslib<span class="op">/</span>releases<span class="op">/</span>download<span class="op">/</span><span class="fl">1.9</span><span class="op">/</span>htslib<span class="op">-</span><span class="fl">1.9</span>.tar.bz2</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>tar <span class="op">-</span>vxjf htslib<span class="op">-</span><span class="fl">1.9</span>.tar.bz2</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="op">%</span>cd htslib<span class="op">-</span><span class="fl">1.9</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>autoreconf <span class="op">-</span>i</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>.<span class="op">/</span>configure</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>make</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>make install</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-27" class="cell" data-execution_count="21">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co">## if you Permission denied error on a mac, try</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>brew install htslib</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<div class="ansi-escaped-output">
<pre><span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Auto-updating Homebrew...</span>

Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with

HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Downloading https://ghcr.io/v2/homebrew/portable-ruby/portable-ruby/blobs/sha256:40e7f5d7514a7e9757facdd39006f7a351d3d7986d3a228be13c8b1c3216727b</span>

######################################################################### 100.0%

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Pouring portable-ruby-3.4.3.arm64_big_sur.bottle.tar.gz</span>

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Auto-updated Homebrew!</span>

Updated 2 taps (homebrew/core and homebrew/cask).

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">New Formulae</span>

bkmr                       newsraft                   tdom

camlpdf                    policy-engine              tmex

elfio                      readerwriterqueue          undercutf1

iccmax                     rofi                       unoserver

infat                      sequoia-chameleon-gnupg

miniflux                   sexpect

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">New Casks</span>

cloudpouch                               font-harmonyos-sans-sc

clover-chord-systems                     font-harmonyos-sans-tc

elemental                                font-noto-serif-dives-akuru

elemental@6                              font-wdxl-lubrifont-jp-n

font-asta-sans                           font-wdxl-lubrifont-sc

font-bizter                              font-wdxl-lubrifont-tc

font-harmonyos-sans                      meru

font-harmonyos-sans-naskh-arabic         sc-menu



You have <span class="ansi-bold">13</span> outdated formulae installed.



<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Downloading https://ghcr.io/v2/homebrew/core/htslib/manifests/1.21</span>

######################################################################### 100.0%

<span class="ansi-green-fg">==&gt;</span> <span class="ansi-bold">Fetching </span><span class="ansi-green-fg ansi-bold">htslib</span>

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Downloading https://ghcr.io/v2/homebrew/core/htslib/blobs/sha256:5666590893b</span>

######################################################################### 100.0%

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Pouring htslib--1.21.arm64_sequoia.bottle.tar.gz</span>

🍺  /opt/homebrew/Cellar/htslib/1.21: 51 files, 5.6MB

<span class="ansi-blue-fg">==&gt;</span> <span class="ansi-bold">Running `brew cleanup htslib`...</span>

Disable this behaviour by setting HOMEBREW_NO_INSTALL_CLEANUP.

Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).
</pre>
</div>
</div>
</div>
</section>
<section id="index-vcf" class="level4">
<h4 class="anchored" data-anchor-id="index-vcf">Index vcf</h4>
<div id="cell-29" class="cell" data-execution_count="22">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>tabix {VCF_FILE}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="define-paths" class="level3">
<h3 class="anchored" data-anchor-id="define-paths">Define paths</h3>
<div id="cell-31" class="cell" data-execution_count="40">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>models_path <span class="op">=</span> CONTENT_DIR <span class="op">+</span> <span class="st">'/data/saved_models/'</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>params_file <span class="op">=</span> GITHUB_DIR <span class="op">+</span> <span class="st">'/borzoi/examples/params_pred.json'</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>targets_file <span class="op">=</span> GITHUB_DIR <span class="op">+</span> <span class="st">'/borzoi/examples/targets_human.txt'</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Load splice site annotation</p>
<div id="cell-33" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="c3e5471a-3ba8-41e8-9ea1-762853e47dcc" data-execution_count="26">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>splice_df <span class="op">=</span> pd.read_csv(CONTENT_DIR<span class="op">+</span><span class="st">'/data/gencode41_basic_protein_splice.csv.gz'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, compression<span class="op">=</span><span class="st">'gzip'</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="st">"len(splice_df) = "</span> <span class="op">+</span> <span class="bu">str</span>(<span class="bu">len</span>(splice_df)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>len(splice_df) = 404837</code></pre>
</div>
</div>
<p>Load transcriptome</p>
<div id="cell-35" class="cell" data-execution_count="29">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>transcriptome <span class="op">=</span> bgene.Transcriptome(CONTENT_DIR <span class="op">+</span> <span class="st">'/data/gencode41_basic_nort.gtf'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="define-functions" class="level2">
<h2 class="anchored" data-anchor-id="define-functions">2. Define functions</h2>
<section id="fastastringextractor" class="level4">
<h4 class="anchored" data-anchor-id="fastastringextractor">FastaStringExtractor</h4>
<p>From reference seq fasta: - Indexes ref sequence - Gets chromosome sizes into a dictionary</p>
<p>Extract function: - gets target chromosome length from kipoiseq interval and chr sizes dictionary - truncates interval if it extends beyond chromosome lengths - extracts sequence</p>
</section>
<section id="make_seq_1hot" class="level4">
<h4 class="anchored" data-anchor-id="make_seq_1hot">make_seq_1hot</h4>
<ul>
<li>one hot encodes given sequence</li>
</ul>
</section>
<section id="predict_tracks" class="level4">
<h4 class="anchored" data-anchor-id="predict_tracks">predict_tracks</h4>
<ul>
<li>makes prediction of ALL tracks given an encoded sequence</li>
</ul>
<div id="cell-38" class="cell" data-execution_count="31">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> FastaStringExtractor:</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, fasta_file):</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>.fasta <span class="op">=</span> pyfaidx.Fasta(fasta_file)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>._chromosome_sizes <span class="op">=</span> {k: <span class="bu">len</span>(v) <span class="cf">for</span> k, v <span class="kw">in</span> <span class="va">self</span>.fasta.items()}</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>    <span class="co">#import pd.Interval as Interval</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> extract(<span class="va">self</span>, interval: Interval, <span class="op">**</span>kwargs) <span class="op">-&gt;</span> <span class="bu">str</span>:</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Truncate interval if it extends beyond the chromosome lengths.</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>        chromosome_length <span class="op">=</span> <span class="va">self</span>._chromosome_sizes[interval.chrom]</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>        trimmed_interval <span class="op">=</span> Interval(interval.chrom,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>                                    <span class="bu">max</span>(interval.start, <span class="dv">0</span>),</span>
<span id="cb27-13"><a href="#cb27-13" aria-hidden="true" tabindex="-1"></a>                                    <span class="bu">min</span>(interval.end, chromosome_length),</span>
<span id="cb27-14"><a href="#cb27-14" aria-hidden="true" tabindex="-1"></a>                                    )</span>
<span id="cb27-15"><a href="#cb27-15" aria-hidden="true" tabindex="-1"></a>        <span class="co"># pyfaidx wants a 1-based interval</span></span>
<span id="cb27-16"><a href="#cb27-16" aria-hidden="true" tabindex="-1"></a>        sequence <span class="op">=</span> <span class="bu">str</span>(<span class="va">self</span>.fasta.get_seq(trimmed_interval.chrom,</span>
<span id="cb27-17"><a href="#cb27-17" aria-hidden="true" tabindex="-1"></a>                                          trimmed_interval.start <span class="op">+</span> <span class="dv">1</span>,</span>
<span id="cb27-18"><a href="#cb27-18" aria-hidden="true" tabindex="-1"></a>                                          trimmed_interval.stop).seq).upper()</span>
<span id="cb27-19"><a href="#cb27-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Fill truncated values with N's.</span></span>
<span id="cb27-20"><a href="#cb27-20" aria-hidden="true" tabindex="-1"></a>        pad_upstream <span class="op">=</span> <span class="st">'N'</span> <span class="op">*</span> <span class="bu">max</span>(<span class="op">-</span>interval.start, <span class="dv">0</span>)</span>
<span id="cb27-21"><a href="#cb27-21" aria-hidden="true" tabindex="-1"></a>        pad_downstream <span class="op">=</span> <span class="st">'N'</span> <span class="op">*</span> <span class="bu">max</span>(interval.end <span class="op">-</span> chromosome_length, <span class="dv">0</span>)</span>
<span id="cb27-22"><a href="#cb27-22" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> pad_upstream <span class="op">+</span> sequence <span class="op">+</span> pad_downstream <span class="co"># returns padded sequence</span></span>
<span id="cb27-23"><a href="#cb27-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-24"><a href="#cb27-24" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> close(<span class="va">self</span>):</span>
<span id="cb27-25"><a href="#cb27-25" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="va">self</span>.fasta.close()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-39" class="cell" data-execution_count="32">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make one-hot coded sequence (modified from borzoi)</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> make_seq_1hot(sequence):</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> dna.dna_1hot(sequence).astype(<span class="st">"float32"</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> predict_tracks(models, sequence_one_hot):</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  predicted_tracks <span class="op">=</span> []</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> fold_ix <span class="kw">in</span> <span class="bu">range</span>(<span class="bu">len</span>(models)):</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>    yh <span class="op">=</span> models[fold_ix](sequence_one_hot[<span class="va">None</span>, ...])[:, <span class="va">None</span>, ...].astype(<span class="st">"float16"</span>)</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    predicted_tracks.append(yh)</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  predicted_tracks <span class="op">=</span> np.concatenate(predicted_tracks, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> predicted_tracks</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="vcf_to_seq_faster" class="level4">
<h4 class="anchored" data-anchor-id="vcf_to_seq_faster">vcf_to_seq_faster</h4>
<ul>
<li>Modifies ref sequence to personal sequence given the vcf</li>
</ul>
<div id="cell-41" class="cell" data-execution_count="33">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> vcf_to_seq_faster(target_interval, individual, vcf_file, fasta_extractor):</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># target_inverval is a kipoiseq interval, e.g. kipoiseq.Interval("chr22", 18118779, 18145669)</span></span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># individual is the id of the individual in the vcf file</span></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>  target_fa <span class="op">=</span> fasta_extractor.extract(target_interval.resize(SEQUENCE_LENGTH))</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>  window_coords <span class="op">=</span> target_interval.resize(SEQUENCE_LENGTH)</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>      <span class="co"># two haplotypes per individual</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>  haplo_1 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>  haplo_2 <span class="op">=</span> <span class="bu">list</span>(target_fa[:])</span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Open the VCF file</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>  vcf_reader <span class="op">=</span> cyvcf2.VCF(vcf_file)</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-14"><a href="#cb29-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Specific genomic region</span></span>
<span id="cb29-15"><a href="#cb29-15" aria-hidden="true" tabindex="-1"></a>  CHROM <span class="op">=</span> window_coords.chrom</span>
<span id="cb29-16"><a href="#cb29-16" aria-hidden="true" tabindex="-1"></a>  START <span class="op">=</span> <span class="bu">max</span>(<span class="dv">1</span>,window_coords.start)</span>
<span id="cb29-17"><a href="#cb29-17" aria-hidden="true" tabindex="-1"></a>  END <span class="op">=</span> <span class="bu">min</span>(window_coords.end, fasta_extractor._chromosome_sizes[CHROM]<span class="op">-</span><span class="dv">1</span>)</span>
<span id="cb29-18"><a href="#cb29-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-19"><a href="#cb29-19" aria-hidden="true" tabindex="-1"></a>  count1 <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-20"><a href="#cb29-20" aria-hidden="true" tabindex="-1"></a>  count2 <span class="op">=</span> <span class="dv">0</span></span>
<span id="cb29-21"><a href="#cb29-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-22"><a href="#cb29-22" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Iterate through variants in the specified region</span></span>
<span id="cb29-23"><a href="#cb29-23" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> variant <span class="kw">in</span> vcf_reader(CHROM <span class="op">+</span> <span class="st">':'</span> <span class="op">+</span> <span class="bu">str</span>(START) <span class="op">+</span> <span class="st">'-'</span> <span class="op">+</span> <span class="bu">str</span>(END)):</span>
<span id="cb29-24"><a href="#cb29-24" aria-hidden="true" tabindex="-1"></a>      <span class="co"># Access the individual's genotype using index (0-based) of the sample</span></span>
<span id="cb29-25"><a href="#cb29-25" aria-hidden="true" tabindex="-1"></a>      individual_index <span class="op">=</span> vcf_reader.samples.index(individual)</span>
<span id="cb29-26"><a href="#cb29-26" aria-hidden="true" tabindex="-1"></a>      genotype <span class="op">=</span> variant.genotypes[individual_index]</span>
<span id="cb29-27"><a href="#cb29-27" aria-hidden="true" tabindex="-1"></a>      ALT<span class="op">=</span>variant.ALT[<span class="dv">0</span>]</span>
<span id="cb29-28"><a href="#cb29-28" aria-hidden="true" tabindex="-1"></a>      REF<span class="op">=</span>variant.REF</span>
<span id="cb29-29"><a href="#cb29-29" aria-hidden="true" tabindex="-1"></a>      POS<span class="op">=</span>variant.POS</span>
<span id="cb29-30"><a href="#cb29-30" aria-hidden="true" tabindex="-1"></a>      <span class="cf">if</span> REF <span class="op">==</span> target_fa[POS <span class="op">-</span> window_coords.start <span class="op">-</span> <span class="dv">1</span>]:</span>
<span id="cb29-31"><a href="#cb29-31" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> genotype[<span class="dv">0</span>] <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb29-32"><a href="#cb29-32" aria-hidden="true" tabindex="-1"></a>          haplo_1[POS <span class="op">-</span> window_coords.start <span class="op">-</span> <span class="dv">1</span>] <span class="op">=</span> ALT</span>
<span id="cb29-33"><a href="#cb29-33" aria-hidden="true" tabindex="-1"></a>          count1 <span class="op">=</span> count1 <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb29-34"><a href="#cb29-34" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> genotype[<span class="dv">1</span>] <span class="op">==</span> <span class="dv">1</span>:</span>
<span id="cb29-35"><a href="#cb29-35" aria-hidden="true" tabindex="-1"></a>          haplo_2[POS <span class="op">-</span> window_coords.start <span class="op">-</span> <span class="dv">1</span>] <span class="op">=</span> ALT</span>
<span id="cb29-36"><a href="#cb29-36" aria-hidden="true" tabindex="-1"></a>          count2 <span class="op">=</span> count2 <span class="op">+</span> <span class="dv">1</span></span>
<span id="cb29-37"><a href="#cb29-37" aria-hidden="true" tabindex="-1"></a>      <span class="cf">else</span>:</span>
<span id="cb29-38"><a href="#cb29-38" aria-hidden="true" tabindex="-1"></a>        <span class="bu">print</span>(<span class="st">"ERROR: REF in vcf "</span><span class="op">+</span> REF <span class="op">+</span> <span class="st">"!= REF from the build"</span> <span class="op">+</span> target_fa[POS <span class="op">-</span> window_coords.start <span class="op">-</span> <span class="dv">1</span>])</span>
<span id="cb29-39"><a href="#cb29-39" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"number of changes haplo1:"</span>)</span>
<span id="cb29-40"><a href="#cb29-40" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(count1)</span>
<span id="cb29-41"><a href="#cb29-41" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(<span class="st">"number of changes haplo2:"</span>)</span>
<span id="cb29-42"><a href="#cb29-42" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(count2)</span>
<span id="cb29-43"><a href="#cb29-43" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> haplo_1, haplo_2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="make_prediction-function" class="level4">
<h4 class="anchored" data-anchor-id="make_prediction-function">make_prediction function</h4>
<div id="cell-43" class="cell" data-execution_count="34">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> make_prediction(gene, interval, haplo, sequence_one_hot, window <span class="op">=</span> <span class="dv">131072</span>): <span class="co"># snp_pos, alt_allele</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> tf.device(<span class="st">'/GPU:0'</span>):</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>        search_gene <span class="op">=</span> gene <span class="co"># gene to predict</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>        resized_interval <span class="op">=</span> interval.resize(SEQUENCE_LENGTH)</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>        start <span class="op">=</span> resized_interval.start</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>        end <span class="op">=</span> resized_interval.end</span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>        center_pos <span class="op">=</span> start <span class="op">+</span> SEQUENCE_LENGTH<span class="op">//</span><span class="dv">2</span> <span class="co"># figure center position</span></span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>        chrom <span class="op">=</span> resized_interval.<span class="bu">chr</span> <span class="co"># chromosome</span></span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Get exon bin range</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>        gene_keys <span class="op">=</span> [gene_key <span class="cf">for</span> gene_key <span class="kw">in</span> transcriptome.genes.keys() <span class="cf">if</span> search_gene <span class="kw">in</span> gene_key]</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>        gene <span class="op">=</span> transcriptome.genes[gene_keys[<span class="dv">0</span>]]</span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Determine output sequence start</span></span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>        seq_out_start <span class="op">=</span> start <span class="op">+</span> seqnn_model.model_strides[<span class="dv">0</span>]<span class="op">*</span>seqnn_model.target_crops[<span class="dv">0</span>]</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>        seq_out_len <span class="op">=</span> seqnn_model.model_strides[<span class="dv">0</span>]<span class="op">*</span>seqnn_model.target_lengths[<span class="dv">0</span>]</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-21"><a href="#cb30-21" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Determine output positions of gene exons</span></span>
<span id="cb30-22"><a href="#cb30-22" aria-hidden="true" tabindex="-1"></a>        gene_slice <span class="op">=</span> gene.output_slice(seq_out_start, seq_out_len, seqnn_model.model_strides[<span class="dv">0</span>], <span class="va">False</span>)</span>
<span id="cb30-23"><a href="#cb30-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb30-24"><a href="#cb30-24" aria-hidden="true" tabindex="-1"></a>        <span class="co">#Make predictions</span></span>
<span id="cb30-25"><a href="#cb30-25" aria-hidden="true" tabindex="-1"></a>        <span class="co">#y_wt</span></span>
<span id="cb30-26"><a href="#cb30-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> predict_tracks(models, sequence_one_hot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="get-indexes-of-tracks-function" class="level4">
<h4 class="anchored" data-anchor-id="get-indexes-of-tracks-function">Get indexes of tracks function</h4>
<div id="cell-45" class="cell" data-execution_count="35">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get indexes of track</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_track_idx(tracks):</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>  track_idx <span class="op">=</span> []</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>  targets_df[<span class="st">'local_index'</span>] <span class="op">=</span> np.arange(<span class="bu">len</span>(targets_df))</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> track <span class="kw">in</span> tracks:</span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>    track_idx.append(targets_df.loc[targets_df[<span class="st">'description'</span>] <span class="op">==</span> track][<span class="st">'local_index'</span>].tolist())</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>  <span class="cf">return</span> track_idx</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="plot-tracks-function" class="level4">
<h4 class="anchored" data-anchor-id="plot-tracks-function">Plot tracks function</h4>
<div id="cell-47" class="cell" data-execution_count="36">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> plot_tracks(prediction, interval, haplo, anno_df<span class="op">=</span><span class="va">None</span>, tracks<span class="op">=</span>[],</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>                log <span class="op">=</span> [<span class="va">False</span>, <span class="va">False</span>, <span class="va">False</span>],</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>                sqrt_scale <span class="op">=</span> [<span class="va">False</span>, <span class="va">False</span>, <span class="va">False</span>]):</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Gene slice not included</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>  bin_size <span class="op">=</span> <span class="dv">32</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a>  pad <span class="op">=</span> <span class="dv">16</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a>  resized_interval <span class="op">=</span> interval.resize(SEQUENCE_LENGTH)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a>  start <span class="op">=</span> resized_interval.start</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a>  end <span class="op">=</span> resized_interval.end</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a>  center_pos <span class="op">=</span> start <span class="op">+</span> SEQUENCE_LENGTH<span class="op">//</span><span class="dv">2</span> <span class="co"># figure center position</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a>  window<span class="op">=</span><span class="dv">131072</span></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a>  plot_start <span class="op">=</span> center_pos <span class="op">-</span> window <span class="op">//</span> <span class="dv">2</span></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>  plot_end <span class="op">=</span> center_pos <span class="op">+</span> window <span class="op">//</span> <span class="dv">2</span></span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>  plot_start_bin <span class="op">=</span> (plot_start <span class="op">-</span> start) <span class="op">//</span> bin_size <span class="op">-</span> pad</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a>  plot_end_bin <span class="op">=</span> (plot_end <span class="op">-</span> start) <span class="op">//</span> bin_size <span class="op">-</span> pad</span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a>  track_idx <span class="op">=</span> get_track_idx(tracks)</span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>  track_scales <span class="op">=</span> [<span class="fl">0.01</span>, <span class="fl">0.01</span>, <span class="fl">0.01</span>]</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a>  track_transforms <span class="op">=</span> [<span class="fl">3.</span><span class="op">/</span><span class="fl">4.</span>, <span class="fl">3.</span><span class="op">/</span><span class="fl">4.</span>, <span class="fl">3.</span><span class="op">/</span><span class="fl">4.</span>]</span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a>  soft_clips <span class="op">=</span> [<span class="fl">384.</span>, <span class="fl">384.</span>, <span class="fl">384.</span>]</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get annotation positions</span></span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a>  anno_poses <span class="op">=</span> []</span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> anno_df <span class="kw">is</span> <span class="kw">not</span> <span class="va">None</span>:</span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>      anno_poses <span class="op">=</span> anno_df.query(</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a>          <span class="st">"chrom == '"</span></span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a>          <span class="op">+</span> interval.<span class="bu">chr</span></span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a>          <span class="op">+</span> <span class="st">"' and position_hg38 &gt;= "</span></span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a>          <span class="op">+</span> <span class="bu">str</span>(plot_start)</span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a>          <span class="op">+</span> <span class="st">" and position_hg38 &lt; "</span></span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a>          <span class="op">+</span> <span class="bu">str</span>(plot_end)</span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a>        )[<span class="st">"position_hg38"</span>].values.tolist()</span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a>  <span class="co"># BIG plot</span></span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a>  fig, axes <span class="op">=</span> plt.subplots(<span class="bu">len</span>(tracks), figsize<span class="op">=</span>(<span class="dv">20</span>, <span class="fl">1.5</span> <span class="op">*</span> <span class="bu">len</span>(tracks)), sharex <span class="op">=</span> <span class="va">True</span>)</span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-39"><a href="#cb32-39" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> ax, track, track_index, track_scale, track_transform, clip_soft <span class="kw">in</span> <span class="bu">zip</span>(axes,</span>
<span id="cb32-40"><a href="#cb32-40" aria-hidden="true" tabindex="-1"></a>        tracks, track_idx, track_scales, track_transforms, soft_clips):</span>
<span id="cb32-41"><a href="#cb32-41" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Plot track densities</span></span>
<span id="cb32-42"><a href="#cb32-42" aria-hidden="true" tabindex="-1"></a>    y <span class="op">=</span> np.array(np.copy(prediction), dtype <span class="op">=</span> np.float32)</span>
<span id="cb32-43"><a href="#cb32-43" aria-hidden="true" tabindex="-1"></a>    y <span class="op">=</span> np.mean(y[..., track_index], axis<span class="op">=</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb32-44"><a href="#cb32-44" aria-hidden="true" tabindex="-1"></a>    y <span class="op">=</span> y[plot_start_bin:plot_end_bin]</span>
<span id="cb32-45"><a href="#cb32-45" aria-hidden="true" tabindex="-1"></a>    signal <span class="op">=</span> np.<span class="bu">max</span>(y)</span>
<span id="cb32-46"><a href="#cb32-46" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> log:</span>
<span id="cb32-47"><a href="#cb32-47" aria-hidden="true" tabindex="-1"></a>      y <span class="op">=</span> np.log2(y <span class="op">+</span> <span class="fl">1.0</span>)</span>
<span id="cb32-48"><a href="#cb32-48" aria-hidden="true" tabindex="-1"></a>    <span class="cf">elif</span> sqrt_scale:</span>
<span id="cb32-49"><a href="#cb32-49" aria-hidden="true" tabindex="-1"></a>      y <span class="op">=</span> np.sqrt(y <span class="op">+</span> <span class="fl">1.0</span>)</span>
<span id="cb32-50"><a href="#cb32-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-51"><a href="#cb32-51" aria-hidden="true" tabindex="-1"></a>    max_y <span class="op">=</span> np.<span class="bu">max</span>(y)</span>
<span id="cb32-52"><a href="#cb32-52" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Max signal for "</span><span class="op">+</span> <span class="bu">str</span>(track) <span class="op">+</span> <span class="st">" = "</span> <span class="op">+</span> <span class="bu">str</span>(<span class="bu">round</span>(signal, <span class="dv">4</span>)))</span>
<span id="cb32-53"><a href="#cb32-53" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"Max transformed signal for "</span><span class="op">+</span> <span class="bu">str</span>(track) <span class="op">+</span> <span class="st">" = "</span> <span class="op">+</span> <span class="bu">str</span>(<span class="bu">round</span>(max_y, <span class="dv">4</span>)))</span>
<span id="cb32-54"><a href="#cb32-54" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"---"</span>)</span>
<span id="cb32-55"><a href="#cb32-55" aria-hidden="true" tabindex="-1"></a>    ax.bar(np.arange(plot_end_bin <span class="op">-</span> plot_start_bin) <span class="op">+</span> plot_start_bin,</span>
<span id="cb32-56"><a href="#cb32-56" aria-hidden="true" tabindex="-1"></a>            y,</span>
<span id="cb32-57"><a href="#cb32-57" aria-hidden="true" tabindex="-1"></a>            width<span class="op">=</span><span class="fl">1.0</span>,</span>
<span id="cb32-58"><a href="#cb32-58" aria-hidden="true" tabindex="-1"></a>            color<span class="op">=</span><span class="st">"green"</span>,</span>
<span id="cb32-59"><a href="#cb32-59" aria-hidden="true" tabindex="-1"></a>            alpha<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb32-60"><a href="#cb32-60" aria-hidden="true" tabindex="-1"></a>            <span class="co"># label="Prediction",)</span></span>
<span id="cb32-61"><a href="#cb32-61" aria-hidden="true" tabindex="-1"></a>    <span class="co"># X axis tick annotations</span></span>
<span id="cb32-62"><a href="#cb32-62" aria-hidden="true" tabindex="-1"></a>    xtick_vals <span class="op">=</span> []</span>
<span id="cb32-63"><a href="#cb32-63" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> _, anno_pos <span class="kw">in</span> <span class="bu">enumerate</span>(anno_poses):</span>
<span id="cb32-64"><a href="#cb32-64" aria-hidden="true" tabindex="-1"></a>      pas_bin <span class="op">=</span> <span class="bu">int</span>((anno_pos <span class="op">-</span> start) <span class="op">//</span> <span class="dv">32</span>) <span class="op">-</span> <span class="dv">16</span></span>
<span id="cb32-65"><a href="#cb32-65" aria-hidden="true" tabindex="-1"></a>      xtick_vals.append(pas_bin)</span>
<span id="cb32-66"><a href="#cb32-66" aria-hidden="true" tabindex="-1"></a>      bin_end <span class="op">=</span> pas_bin <span class="op">+</span> <span class="dv">3</span> <span class="op">-</span> <span class="fl">0.5</span></span>
<span id="cb32-67"><a href="#cb32-67" aria-hidden="true" tabindex="-1"></a>      bin_start <span class="op">=</span> bin_end <span class="op">-</span> <span class="dv">5</span></span>
<span id="cb32-68"><a href="#cb32-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-69"><a href="#cb32-69" aria-hidden="true" tabindex="-1"></a>      ax.axvline(x<span class="op">=</span>pas_bin,</span>
<span id="cb32-70"><a href="#cb32-70" aria-hidden="true" tabindex="-1"></a>                color<span class="op">=</span><span class="st">"cyan"</span>,</span>
<span id="cb32-71"><a href="#cb32-71" aria-hidden="true" tabindex="-1"></a>                linewidth<span class="op">=</span><span class="dv">2</span>,</span>
<span id="cb32-72"><a href="#cb32-72" aria-hidden="true" tabindex="-1"></a>                alpha<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb32-73"><a href="#cb32-73" aria-hidden="true" tabindex="-1"></a>                linestyle<span class="op">=</span><span class="st">"-"</span>,</span>
<span id="cb32-74"><a href="#cb32-74" aria-hidden="true" tabindex="-1"></a>                zorder<span class="op">=-</span><span class="dv">1</span>,)</span>
<span id="cb32-75"><a href="#cb32-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-76"><a href="#cb32-76" aria-hidden="true" tabindex="-1"></a>    plt.xlim(plot_start_bin, plot_end_bin <span class="op">-</span> <span class="dv">1</span>)</span>
<span id="cb32-77"><a href="#cb32-77" aria-hidden="true" tabindex="-1"></a>    plt.xticks([], [])</span>
<span id="cb32-78"><a href="#cb32-78" aria-hidden="true" tabindex="-1"></a>    plt.yticks([], [])</span>
<span id="cb32-79"><a href="#cb32-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-80"><a href="#cb32-80" aria-hidden="true" tabindex="-1"></a>    ax.<span class="bu">set</span>(ylabel <span class="op">=</span> <span class="st">"Signal(log)"</span> <span class="cf">if</span> log <span class="cf">else</span> <span class="st">"Signal"</span>)<span class="co"># , fontsize = 8)</span></span>
<span id="cb32-81"><a href="#cb32-81" aria-hidden="true" tabindex="-1"></a>    ax.set_title(<span class="st">"Track: "</span> <span class="op">+</span> <span class="bu">str</span>(track)<span class="op">+</span> <span class="st">" in "</span><span class="op">+</span> <span class="bu">str</span>(haplo), fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb32-82"><a href="#cb32-82" aria-hidden="true" tabindex="-1"></a>    <span class="co"># plt.legend(fontsize=8)</span></span>
<span id="cb32-83"><a href="#cb32-83" aria-hidden="true" tabindex="-1"></a>    plt.tight_layout()</span>
<span id="cb32-84"><a href="#cb32-84" aria-hidden="true" tabindex="-1"></a>  ax.set_xlabel(</span>
<span id="cb32-85"><a href="#cb32-85" aria-hidden="true" tabindex="-1"></a>            interval.<span class="bu">chr</span></span>
<span id="cb32-86"><a href="#cb32-86" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="st">":"</span></span>
<span id="cb32-87"><a href="#cb32-87" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="bu">str</span>(plot_start)</span>
<span id="cb32-88"><a href="#cb32-88" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="st">"-"</span></span>
<span id="cb32-89"><a href="#cb32-89" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="bu">str</span>(plot_end)</span>
<span id="cb32-90"><a href="#cb32-90" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="st">" ("</span></span>
<span id="cb32-91"><a href="#cb32-91" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="bu">str</span>(window)</span>
<span id="cb32-92"><a href="#cb32-92" aria-hidden="true" tabindex="-1"></a>            <span class="op">+</span> <span class="st">"bp window)"</span>)</span>
<span id="cb32-93"><a href="#cb32-93" aria-hidden="true" tabindex="-1"></a>            <span class="co"># fontsize=8,)</span></span>
<span id="cb32-94"><a href="#cb32-94" aria-hidden="true" tabindex="-1"></a>  plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="extract-fasta" class="level3">
<h3 class="anchored" data-anchor-id="extract-fasta">Extract fasta</h3>
<div id="cell-49" class="cell" data-execution_count="37">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>fasta_extractor <span class="op">=</span> FastaStringExtractor(fasta_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="model-configuration" class="level2">
<h2 class="anchored" data-anchor-id="model-configuration">3.Model configuration</h2>
<div id="cell-51" class="cell" data-execution_count="43">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>SEQUENCE_LENGTH <span class="op">=</span> <span class="dv">524288</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>n_folds <span class="op">=</span> <span class="dv">4</span>       <span class="co">#To use only one model fold, change to 'n_folds = 1'</span></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>rc <span class="op">=</span> <span class="va">True</span>         <span class="co">#Average across reverse-complement prediction</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Read model parameters</span></span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a><span class="cf">with</span> <span class="bu">open</span>(params_file) <span class="im">as</span> params_open :</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>    params <span class="op">=</span> json.load(params_open)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>    params_model <span class="op">=</span> params[<span class="st">'model'</span>]</span>
<span id="cb34-12"><a href="#cb34-12" aria-hidden="true" tabindex="-1"></a>    params_train <span class="op">=</span> params[<span class="st">'train'</span>]</span>
<span id="cb34-13"><a href="#cb34-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-14"><a href="#cb34-14" aria-hidden="true" tabindex="-1"></a><span class="co">#Read targets</span></span>
<span id="cb34-15"><a href="#cb34-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-16"><a href="#cb34-16" aria-hidden="true" tabindex="-1"></a>targets_df <span class="op">=</span> pd.read_csv(targets_file, index_col<span class="op">=</span><span class="dv">0</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb34-17"><a href="#cb34-17" aria-hidden="true" tabindex="-1"></a>target_index <span class="op">=</span> targets_df.index</span>
<span id="cb34-18"><a href="#cb34-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-19"><a href="#cb34-19" aria-hidden="true" tabindex="-1"></a><span class="co">#Create local index of strand_pair (relative to sliced targets)</span></span>
<span id="cb34-20"><a href="#cb34-20" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> rc :</span>
<span id="cb34-21"><a href="#cb34-21" aria-hidden="true" tabindex="-1"></a>    strand_pair <span class="op">=</span> targets_df.strand_pair</span>
<span id="cb34-22"><a href="#cb34-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-23"><a href="#cb34-23" aria-hidden="true" tabindex="-1"></a>    target_slice_dict <span class="op">=</span> {ix : i <span class="cf">for</span> i, ix <span class="kw">in</span> <span class="bu">enumerate</span>(target_index.values.tolist())}</span>
<span id="cb34-24"><a href="#cb34-24" aria-hidden="true" tabindex="-1"></a>    slice_pair <span class="op">=</span> np.array([</span>
<span id="cb34-25"><a href="#cb34-25" aria-hidden="true" tabindex="-1"></a>        target_slice_dict[ix] <span class="cf">if</span> ix <span class="kw">in</span> target_slice_dict <span class="cf">else</span> ix <span class="cf">for</span> ix <span class="kw">in</span> strand_pair.values.tolist()</span>
<span id="cb34-26"><a href="#cb34-26" aria-hidden="true" tabindex="-1"></a>    ], dtype<span class="op">=</span><span class="st">'int32'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-52" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="20540e23-7311-470a-989d-d48c91343a33" data-execution_count="47">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Initialize model ensemble</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>models <span class="op">=</span> []</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> fold_ix <span class="kw">in</span> <span class="bu">range</span>(n_folds) :</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>    model_file <span class="op">=</span> models_path <span class="op">+</span> <span class="st">"f"</span> <span class="op">+</span> <span class="bu">str</span>(fold_ix) <span class="op">+</span> <span class="st">"/model0_best.h5"</span></span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a>    seqnn_model <span class="op">=</span> seqnn.SeqNN(params_model)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a>    seqnn_model.restore(model_file, <span class="dv">0</span>)</span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a>    seqnn_model.build_slice(target_index)</span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> rc :</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>        seqnn_model.strand_pair.append(slice_pair)</span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>    <span class="co">#seqnn_model.build_ensemble(rc, '0')</span></span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>    seqnn_model.build_ensemble(rc, [<span class="dv">0</span>])</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a>    models.append(seqnn_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="build-sequence" class="level2">
<h2 class="anchored" data-anchor-id="build-sequence">4. Build sequence</h2>
<div id="cell-54" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="11e7a561-6077-444d-f5c6-93cc6aad15b2" data-execution_count="49">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># read VCFs and encode haplotypes</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>CHROM <span class="op">=</span> <span class="st">'chr22'</span></span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>vcf_file <span class="op">=</span> CONTENT_DIR <span class="op">+</span> <span class="st">"/data/ALL."</span> <span class="op">+</span> CHROM <span class="op">+</span> <span class="st">".shapeit2_integrated_SNPs_v2a_27022019.GRCh38.phased.gz"</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>target_interval <span class="op">=</span> kipoiseq.Interval(CHROM, <span class="dv">18118779</span>, <span class="dv">18145669</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>haplo1, haplo2 <span class="op">=</span> vcf_to_seq_faster(target_interval, <span class="st">'HG00097'</span>, vcf_file, fasta_extractor)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>haplo0 <span class="op">=</span> fasta_extractor.extract(target_interval.resize(SEQUENCE_LENGTH)) <span class="co"># ref seq</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="co"># removed extra dimension as input layer is of shape=(None, 524288, 4)</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>haplo1_enc <span class="op">=</span> make_seq_1hot(<span class="st">""</span>.join(haplo1))</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>haplo2_enc <span class="op">=</span> make_seq_1hot(<span class="st">""</span>.join(haplo2))</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>haplo0_enc <span class="op">=</span> make_seq_1hot(<span class="st">""</span>.join(haplo0))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>number of changes haplo1:
414
number of changes haplo2:
133</code></pre>
</div>
</div>
</section>
<section id="predict" class="level2">
<h2 class="anchored" data-anchor-id="predict">5. Predict</h2>
<p>ENSG00000099968</p>
<div id="cell-57" class="cell" data-execution_count="50">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>sequences <span class="op">=</span> [haplo0_enc, haplo1_enc, haplo2_enc]</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>my_tracks <span class="op">=</span> [<span class="st">'CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_'</span>,</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                          <span class="st">'DNASE:CD14-positive monocyte female'</span>,</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>                          <span class="st">'RNA:blood'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div id="cell-58" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;,&quot;height&quot;:1000}}" data-outputid="f318a409-356c-4c44-de13-3e88653b7e83" data-execution_count="51">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> h <span class="kw">in</span> <span class="bu">range</span>(<span class="dv">0</span>, <span class="bu">len</span>(sequences)):</span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  haplotypes <span class="op">=</span> [<span class="st">"Reference"</span>, <span class="st">"Haplotype 1"</span>, <span class="st">"Haplotype 2"</span>]</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  prediction <span class="op">=</span> make_prediction(gene <span class="op">=</span> <span class="st">'ENSG00000099968'</span>, interval <span class="op">=</span> target_interval,</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>                               haplo <span class="op">=</span> haplotypes[h], sequence_one_hot<span class="op">=</span>sequences[h])</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="bu">print</span>(haplotypes[h])</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>  plot_tracks(prediction, interval <span class="op">=</span> target_interval, haplo <span class="op">=</span> haplotypes[h],</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>              tracks <span class="op">=</span> my_tracks, log <span class="op">=</span> [<span class="va">True</span>, <span class="va">True</span>, <span class="va">False</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Reference
Max signal for CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_ = 38.1346
Max transformed signal for CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_ = 5.2904
---
Max signal for DNASE:CD14-positive monocyte female = 11.8511
Max transformed signal for DNASE:CD14-positive monocyte female = 3.6838
---
Max signal for RNA:blood = 2.7051
Max transformed signal for RNA:blood = 1.8895
---</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="borzoi_personal_gen_prediction-jupyter_files/figure-html/cell-33-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Haplotype 1
Max signal for CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_ = 39.2737
Max transformed signal for CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_ = 5.3318
---
Max signal for DNASE:CD14-positive monocyte female = 11.8428
Max transformed signal for DNASE:CD14-positive monocyte female = 3.6829
---
Max signal for RNA:blood = 2.689
Max transformed signal for RNA:blood = 1.8832
---</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="borzoi_personal_gen_prediction-jupyter_files/figure-html/cell-33-output-4.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>Haplotype 2
Max signal for CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_ = 38.5659
Max transformed signal for CAGE:B lymphoblastoid cell line: GM12878 ENCODE, biol_ = 5.3062
---
Max signal for DNASE:CD14-positive monocyte female = 11.8301
Max transformed signal for DNASE:CD14-positive monocyte female = 3.6815
---
Max signal for RNA:blood = 2.7414
Max transformed signal for RNA:blood = 1.9036
---</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="borzoi_personal_gen_prediction-jupyter_files/figure-html/cell-33-output-6.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-59" class="cell" data-quarto-private-1="{&quot;key&quot;:&quot;colab&quot;,&quot;value&quot;:{&quot;base_uri&quot;:&quot;https://localhost:8080/&quot;}}" data-outputid="495870e7-f0fd-4b12-9ccf-b53dc0ab7c86" data-execution_count="52">
<details class="code-fold">
<summary>Show the code</summary>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>prediction.shape</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-display" data-execution_count="52">
<pre><code>(1, 4, 16352, 7611)</code></pre>
</div>
</div>


</section>

<p>© <a href="https://hakyimlab.org">HakyImLab and Listed Authors</a> - <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0 License</a></p></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>