<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.42">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Preprocessing and clustering – GENE 46100</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-2f5df379a58b258e96c21c0638c20c03.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap-dcc7d9e706e4adfbab8e3d6c7c60bab2.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


<link rel="stylesheet" href="../../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">GENE 46100</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#quality-control" id="toc-quality-control" class="nav-link active" data-scroll-target="#quality-control">Quality Control</a>
  <ul class="collapse">
  <li><a href="#doublet-detection" id="toc-doublet-detection" class="nav-link" data-scroll-target="#doublet-detection">Doublet detection</a></li>
  </ul></li>
  <li><a href="#normalization" id="toc-normalization" class="nav-link" data-scroll-target="#normalization">Normalization</a></li>
  <li><a href="#feature-selection" id="toc-feature-selection" class="nav-link" data-scroll-target="#feature-selection">Feature selection</a></li>
  <li><a href="#dimensionality-reduction" id="toc-dimensionality-reduction" class="nav-link" data-scroll-target="#dimensionality-reduction">Dimensionality Reduction</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization">Visualization</a></li>
  <li><a href="#clustering" id="toc-clustering" class="nav-link" data-scroll-target="#clustering">Clustering</a></li>
  <li><a href="#re-assess-quality-control-and-cell-filtering" id="toc-re-assess-quality-control-and-cell-filtering" class="nav-link" data-scroll-target="#re-assess-quality-control-and-cell-filtering">Re-assess quality control and cell filtering</a></li>
  <li><a href="#cell-type-annotation" id="toc-cell-type-annotation" class="nav-link" data-scroll-target="#cell-type-annotation">Cell-type annotation</a>
  <ul class="collapse">
  <li><a href="#marker-gene-set" id="toc-marker-gene-set" class="nav-link" data-scroll-target="#marker-gene-set">Marker gene set</a></li>
  <li><a href="#automatic-label-prediction" id="toc-automatic-label-prediction" class="nav-link" data-scroll-target="#automatic-label-prediction">Automatic label prediction</a></li>
  <li><a href="#annotation-with-enrichment-analysis" id="toc-annotation-with-enrichment-analysis" class="nav-link" data-scroll-target="#annotation-with-enrichment-analysis">Annotation with enrichment analysis</a></li>
  <li><a href="#differentially-expressed-genes-as-markers" id="toc-differentially-expressed-genes-as-markers" class="nav-link" data-scroll-target="#differentially-expressed-genes-as-markers">Differentially-expressed Genes as Markers</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Preprocessing and clustering</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div id="f1990713-5968-4fc7-aeb6-5a842606fd8c" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> anndata <span class="im">as</span> ad</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pooch</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> scanpy <span class="im">as</span> sc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="161629d6-b1b6-4faa-8bca-01a841e0ec10" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>sc.settings.set_figure_params(dpi<span class="op">=</span><span class="dv">50</span>, facecolor<span class="op">=</span><span class="st">"white"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The data used in this basic preprocessing and clustering tutorial was collected from bone marrow mononuclear cells of healthy human donors and was part of <a href="https://openproblems.bio/competitions/neurips_2021/">openproblem’s NeurIPS 2021 benchmarking dataset</a> {cite}<code>luecken2021</code>. The samples used in this tutorial were measured using the 10X Multiome Gene Expression and Chromatin Accessability kit.</p>
<p>We are reading in the count matrix into an <a href="https://anndata.readthedocs.io/en/latest/tutorials/notebooks/getting-started.html">AnnData</a> object, which holds many slots for annotations and different representations of the data.</p>
<div id="f0448e97" class="cell" data-execution_count="3">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>EXAMPLE_DATA <span class="op">=</span> pooch.create(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>    path<span class="op">=</span>pooch.os_cache(<span class="st">"scverse_tutorials"</span>),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    base_url<span class="op">=</span><span class="st">"doi:10.6084/m9.figshare.22716739.v1/"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>EXAMPLE_DATA.load_registry_from_doi()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="9bdf7f10-e9b9-4929-a7e2-7ac373d769d1" class="cell" data-execution_count="4">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>samples <span class="op">=</span> {</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"s1d1"</span>: <span class="st">"s1d1_filtered_feature_bc_matrix.h5"</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"s1d3"</span>: <span class="st">"s1d3_filtered_feature_bc_matrix.h5"</span>,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>adatas <span class="op">=</span> {}</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> sample_id, filename <span class="kw">in</span> samples.items():</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    path <span class="op">=</span> EXAMPLE_DATA.fetch(filename)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    sample_adata <span class="op">=</span> sc.read_10x_h5(path)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>    sample_adata.var_names_make_unique()</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    adatas[sample_id] <span class="op">=</span> sample_adata</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> ad.concat(adatas, label<span class="op">=</span><span class="st">"sample"</span>)</span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>adata.obs_names_make_unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading file 's1d1_filtered_feature_bc_matrix.h5' from 'doi:10.6084/m9.figshare.22716739.v1/s1d1_filtered_feature_bc_matrix.h5' to '/home/runner/.cache/scverse_tutorials'.
/home/runner/miniconda3/envs/tutorials/lib/python3.12/site-packages/anndata/_core/anndata.py:1756: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates("var")
/home/runner/miniconda3/envs/tutorials/lib/python3.12/site-packages/anndata/_core/anndata.py:1756: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates("var")
Downloading file 's1d3_filtered_feature_bc_matrix.h5' from 'doi:10.6084/m9.figshare.22716739.v1/s1d3_filtered_feature_bc_matrix.h5' to '/home/runner/.cache/scverse_tutorials'.
/home/runner/miniconda3/envs/tutorials/lib/python3.12/site-packages/anndata/_core/anndata.py:1756: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates("var")
/home/runner/miniconda3/envs/tutorials/lib/python3.12/site-packages/anndata/_core/anndata.py:1756: UserWarning: Variable names are not unique. To make them unique, call `.var_names_make_unique`.
  utils.warn_names_duplicates("var")
/home/runner/miniconda3/envs/tutorials/lib/python3.12/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.
  utils.warn_names_duplicates("obs")</code></pre>
</div>
</div>
<p>The data contains 8,785 cells and 36,601 measured genes. This tutorial includes a basic preprocessing and clustering workflow.</p>
<section id="quality-control" class="level2">
<h2 class="anchored" data-anchor-id="quality-control">Quality Control</h2>
<p>The scanpy function {func}<code>~scanpy.pp.calculate_qc_metrics</code> calculates common quality control (QC) metrics, which are largely based on <code>calculateQCMetrics</code> from scater {cite}<code>McCarthy2017</code>. One can pass specific gene population to {func}<code>~scanpy.pp.calculate_qc_metrics</code> in order to calculate proportions of counts for these populations. Mitochondrial, ribosomal and hemoglobin genes are defined by distinct prefixes as listed below.</p>
<div id="75df4b20-29ca-4e12-a878-94698a1b67d2" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># mitochondrial genes</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">"mt"</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith(<span class="st">"MT-"</span>)  <span class="co"># "MT-" for human, "Mt-" for mouse</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="co"># ribosomal genes</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">"ribo"</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.startswith((<span class="st">"RPS"</span>, <span class="st">"RPL"</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="co"># hemoglobin genes</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>adata.var[<span class="st">"hb"</span>] <span class="op">=</span> adata.var_names.<span class="bu">str</span>.contains(<span class="st">"^HB[^(P)]"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="efd36ca1-75c7-4716-8766-a5571951537b" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>sc.pp.calculate_qc_metrics(adata, qc_vars<span class="op">=</span>[<span class="st">"mt"</span>, <span class="st">"ribo"</span>, <span class="st">"hb"</span>], inplace<span class="op">=</span><span class="va">True</span>, log1p<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One can now inspect violin plots of some of the computed QC metrics:</p>
<ul>
<li>the number of genes expressed in the count matrix</li>
<li>the total counts per cell</li>
<li>the percentage of counts in mitochondrial genes</li>
</ul>
<div id="8f8e3e58-a05a-4fd1-ab77-f3b14232dbf8" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, [<span class="st">"n_genes_by_counts"</span>, <span class="st">"total_counts"</span>, <span class="st">"pct_counts_mt"</span>], jitter<span class="op">=</span><span class="fl">0.4</span>, multi_panel<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-8-output-1.png" width="751" height="238" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Additionally, it is useful to consider QC metrics jointly by inspecting a scatter plot colored by <code>pct_counts_mt</code>.</p>
<div id="6ddfd195-10cd-419d-b984-39bc29dfe56b" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>sc.pl.scatter(adata, <span class="st">"total_counts"</span>, <span class="st">"n_genes_by_counts"</span>, color<span class="op">=</span><span class="st">"pct_counts_mt"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-9-output-1.png" width="214" height="200" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Based on the QC metric plots, one could now remove cells that have too many mitochondrial genes expressed or too many total counts by setting manual or automatic thresholds. However, it proved to be beneficial to apply a very permissive filtering strategy in the beginning for your single-cell analysis and filter low quality cells during clustering or revisit the filtering again at a later point. We therefore now only filter cells with less than 100 genes expressed and genes that are detected in less than 3 cells.</p>
<p>Additionally, it is important to note that for datasets with multiple batches, quality control should be performed for each sample individually as quality control thresholds can very substantially between batches.</p>
<div id="61508aa7-31f1-403c-96c2-09460427bdae" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_cells(adata, min_genes<span class="op">=</span><span class="dv">100</span>)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>sc.pp.filter_genes(adata, min_cells<span class="op">=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="doublet-detection" class="level3">
<h3 class="anchored" data-anchor-id="doublet-detection">Doublet detection</h3>
<p>As a next step, we run a doublet detection algorithm. Identifying doublets is crucial as they can lead to misclassifications or distortions in downstream analysis steps. Scanpy contains the doublet detection method Scrublet {cite}<code>Wolock2019</code>. Scrublet predicts cell doublets using a nearest-neighbor classifier of observed transcriptomes and simulated doublets. {func}<code>scanpy.pp.scrublet</code> adds <code>doublet_score</code> and <code>predicted_doublet</code> to <code>.obs</code>. One can now either filter directly on <code>predicted_doublet</code> or use the <code>doublet_score</code> later during clustering to filter clusters with high doublet scores.</p>
<div id="2bc52329-32e3-4c04-a5a6-be4184391219" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>sc.pp.scrublet(adata, batch_key<span class="op">=</span><span class="st">"sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="{seealso}">
<p>Alternative methods for doublet detection within the scverse ecosystem are <a href="https://github.com/JonathanShor/DoubletDetection">DoubletDetection</a> and <a href="https://docs.scvi-tools.org/en/stable/user_guide/models/solo.html">SOLO</a>. You can read more about these in the <a href="https://www.sc-best-practices.org/preprocessing_visualization/quality_control.html#doublet-detection">Doublet Detection chapter</a> of Single Cell Best Practices.</p>
</div>
</section>
</section>
<section id="normalization" class="level2">
<h2 class="anchored" data-anchor-id="normalization">Normalization</h2>
<p>The next preprocessing step is normalization. A common approach is count depth scaling with subsequent log plus one (log1p) transformation. Count depth scaling normalizes the data to a “size factor” such as the median count depth in the dataset, ten thousand (CP10k) or one million (CPM, counts per million). The size factor for count depth scaling can be controlled via <code>target_sum</code> in <code>pp.normalize_total</code>. We are applying median count depth normalization with log1p transformation (AKA log1PF).</p>
<div id="7fe44a18" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Saving count data</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>adata.layers[<span class="st">"counts"</span>] <span class="op">=</span> adata.X.copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="053aed6d-f793-442d-afa8-46def6054b23" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Normalizing to median total counts</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>sc.pp.normalize_total(adata)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Logarithmize the data:</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>sc.pp.log1p(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="feature-selection" class="level2">
<h2 class="anchored" data-anchor-id="feature-selection">Feature selection</h2>
<p>As a next step, we want to reduce the dimensionality of the dataset and only include the most informative genes. This step is commonly known as feature selection. The scanpy function <code>pp.highly_variable_genes</code> annotates highly variable genes by reproducing the implementations of Seurat {cite}<code>Satija2015</code>, Cell Ranger {cite}<code>Zheng2017</code>, and Seurat v3 {cite}<code>stuart2019comprehensive</code> depending on the chosen <code>flavor</code>.</p>
<div id="6241df3f-a23b-4f06-8c4e-a2de39200f05" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>sc.pp.highly_variable_genes(adata, n_top_genes<span class="op">=</span><span class="dv">2000</span>, batch_key<span class="op">=</span><span class="st">"sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="654984e9-e511-4e32-b025-054d41a1aabd" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>sc.pl.highly_variable_genes(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-15-output-1.png" width="362" height="188" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="dimensionality-reduction" class="level2">
<h2 class="anchored" data-anchor-id="dimensionality-reduction">Dimensionality Reduction</h2>
<p>Reduce the dimensionality of the data by running principal component analysis (PCA), which reveals the main axes of variation and denoises the data.</p>
<div id="5a4d36d8-3238-47a5-8950-049fd0473205" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>sc.tl.pca(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let us inspect the contribution of single PCs to the total variance in the data. This gives us information about how many PCs we should consider in order to compute the neighborhood relations of cells, e.g.&nbsp;used in the clustering function {func}<code>~scanpy.tl.leiden</code> or {func}<code>~scanpy.tl.tsne</code>. In our experience, often a rough estimate of the number of PCs does fine.</p>
<div id="b38f1a27-6c1b-4160-b28e-e172bb325d4d" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>sc.pl.pca_variance_ratio(adata, n_pcs<span class="op">=</span><span class="dv">50</span>, log<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-17-output-1.png" width="187" height="202" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="visualization" class="level2">
<h2 class="anchored" data-anchor-id="visualization">Visualization</h2>
<p>Let us compute the neighborhood graph of cells using the PCA representation of the data matrix.</p>
<div id="ce3adfa7-0be0-4232-88fa-e0540c5c3319" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>sc.pp.neighbors(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>/home/runner/miniconda3/envs/tutorials/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm</code></pre>
</div>
</div>
<p>We suggest embedding the graph in two dimensions using UMAP (McInnes et al., 2018), see below.</p>
<div id="0d0151a0-ce62-40e6-ba1b-2f4505b6b4b4" class="cell" data-execution_count="18">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>sc.tl.umap(adata)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can now visualize the UMAP according to the <code>sample</code>.</p>
<div id="74aba38e-9c07-484b-a222-5f2ec108e7b2" class="cell" data-execution_count="19">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span><span class="st">"sample"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-20-output-1.png" width="219" height="185" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Even though the data considered in this tutorial includes two different samples, we only observe a minor batch effect and we can continue with clustering and annotation of our data.</p>
<p>If you inspect batch effects in your UMAP it can be beneficial to integrate across samples and perform batch correction/integration.</p>
</section>
<section id="clustering" class="level2">
<h2 class="anchored" data-anchor-id="clustering">Clustering</h2>
<p>As with Seurat and many other frameworks, we recommend the Leiden graph-clustering method (community detection based on optimizing modularity) {cite}<code>traag2019louvain</code>. Note that Leiden clustering directly clusters the neighborhood graph of cells, which we already computed in the previous section.</p>
<div id="11bb19be-3df1-4a17-ad84-16b50a9348ea" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, flavor<span class="op">=</span><span class="st">"igraph"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="e616f3f6-69f8-4c66-ac00-48d328dcf755" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="st">"leiden"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-22-output-1.png" width="246" height="187" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="re-assess-quality-control-and-cell-filtering" class="level2">
<h2 class="anchored" data-anchor-id="re-assess-quality-control-and-cell-filtering">Re-assess quality control and cell filtering</h2>
<p>As indicated before, we will now re-assess our filtering strategy by visualizing different QC metrics using UMAP.</p>
<div id="c1818fbe-0e7d-467f-bdd1-3b690c32ad26" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"predicted_doublet"</span>] <span class="op">=</span> adata.obs[<span class="st">"predicted_doublet"</span>].astype(<span class="st">"category"</span>)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"leiden"</span>, <span class="st">"predicted_doublet"</span>, <span class="st">"doublet_score"</span>],</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="co"># increase horizontal space between panels</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>    wspace<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-23-output-1.png" width="750" height="188" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can now subset the AnnData object to exclude cells predicted as doublets:</p>
<div id="bb24ec51" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> adata[<span class="op">~</span>adata.obs[<span class="st">"predicted_doublet"</span>].to_numpy()].copy()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="91ad2b94-ab9b-48a3-aa0c-551203d3cf91" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>    adata, color<span class="op">=</span>[<span class="st">"leiden"</span>, <span class="st">"log1p_total_counts"</span>, <span class="st">"pct_counts_mt"</span>, <span class="st">"log1p_n_genes_by_counts"</span>], wspace<span class="op">=</span><span class="fl">0.5</span>, ncols<span class="op">=</span><span class="dv">2</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-25-output-1.png" width="506" height="366" class="figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="cell-type-annotation" class="level2">
<h2 class="anchored" data-anchor-id="cell-type-annotation">Cell-type annotation</h2>
<div id="acd638c1-a8aa-4dc8-80e5-7827812c82d7" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> celltypist <span class="im">as</span> ct</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> decoupler <span class="im">as</span> dc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We have now reached a point where we have obtained a set of cells with decent quality, and we can proceed to their annotation to known cell types. Typically, this is done using genes that are exclusively expressed by a given cell type, or in other words these genes are the marker genes of the cell types, and are thus used to distinguish the heterogeneous groups of cells in our data. Previous efforts have collected and curated various marker genes into available resources, such as <a href="http://bio-bigdata.hrbmu.edu.cn/CellMarker/">CellMarker</a>, <a href="http://bio.liclab.net/TF-Marker/">TF-Marker</a>, and <a href="https://panglaodb.se/">PanglaoDB</a>.</p>
<p>Commonly and classically, cell type annotation uses those marker genes subsequent to the grouping of the cells into clusters. So, let’s generate a set of clustering solutions which we can then use to annotate our cell types. Here, we will use the Leiden clustering algorithm which will extract cell communities from our nearest neighbours graph.</p>
<div id="148a1050-a073-4d82-b568-3cb943f9fd34" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, flavor<span class="op">=</span><span class="st">"igraph"</span>, key_added<span class="op">=</span><span class="st">"leiden_res0_02"</span>, resolution<span class="op">=</span><span class="fl">0.02</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, flavor<span class="op">=</span><span class="st">"igraph"</span>, key_added<span class="op">=</span><span class="st">"leiden_res0_5"</span>, resolution<span class="op">=</span><span class="fl">0.5</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>sc.tl.leiden(adata, flavor<span class="op">=</span><span class="st">"igraph"</span>, key_added<span class="op">=</span><span class="st">"leiden_res2"</span>, resolution<span class="op">=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notably, the number of clusters that we define is largely arbitrary, and so is the <code>resolution</code> parameter that we use to control for it. As such, the number of clusters is ultimately bound to the stable and biologically-meaningful groups that we can ultimately distringuish, typically done by experts in the corresponding field or by using expert-curated prior knowledge in the form of markers.</p>
<div id="b1275cb4-9f1e-4f37-8c90-eb5d0abd679c" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    adata,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[<span class="st">"leiden_res0_02"</span>, <span class="st">"leiden_res0_5"</span>, <span class="st">"leiden_res2"</span>],</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>    legend_loc<span class="op">=</span><span class="st">"on data"</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-28-output-1.png" width="584" height="187" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Though UMAPs should not be over-interpreted, here we can already see that in the highest resolution our data is over-clustered, while the lowest resolution is likely grouping cells which belong to distinct cell identities.</p>
<section id="marker-gene-set" class="level3">
<h3 class="anchored" data-anchor-id="marker-gene-set">Marker gene set</h3>
<p>Let’s define a set of marker genes for the main cell types that we expect to see in this dataset. These were adapted from <a href="https://www.sc-best-practices.org/cellular_structure/annotation.html">Single Cell Best Practices annotation chapter</a>, for a more detailed overview and best practices in cell type annotation, we refer the user to it.</p>
<div id="c4b1f83e-c4e2-48e4-954f-72abac5b5b21" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>marker_genes <span class="op">=</span> {</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD14+ Mono"</span>: [<span class="st">"FCN1"</span>, <span class="st">"CD14"</span>],</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD16+ Mono"</span>: [<span class="st">"TCF7L2"</span>, <span class="st">"FCGR3A"</span>, <span class="st">"LYN"</span>],</span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Note: DMXL2 should be negative</span></span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"cDC2"</span>: [<span class="st">"CST3"</span>, <span class="st">"COTL1"</span>, <span class="st">"LYZ"</span>, <span class="st">"DMXL2"</span>, <span class="st">"CLEC10A"</span>, <span class="st">"FCER1A"</span>],</span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Erythroblast"</span>: [<span class="st">"MKI67"</span>, <span class="st">"HBA1"</span>, <span class="st">"HBB"</span>],</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Note HBM and GYPA are negative markers</span></span>
<span id="cb30-8"><a href="#cb30-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Proerythroblast"</span>: [<span class="st">"CDK6"</span>, <span class="st">"SYNGR1"</span>, <span class="st">"HBM"</span>, <span class="st">"GYPA"</span>],</span>
<span id="cb30-9"><a href="#cb30-9" aria-hidden="true" tabindex="-1"></a>    <span class="st">"NK"</span>: [<span class="st">"GNLY"</span>, <span class="st">"NKG7"</span>, <span class="st">"CD247"</span>, <span class="st">"FCER1G"</span>, <span class="st">"TYROBP"</span>, <span class="st">"KLRG1"</span>, <span class="st">"FCGR3A"</span>],</span>
<span id="cb30-10"><a href="#cb30-10" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ILC"</span>: [<span class="st">"ID2"</span>, <span class="st">"PLCG2"</span>, <span class="st">"GNLY"</span>, <span class="st">"SYNE1"</span>],</span>
<span id="cb30-11"><a href="#cb30-11" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Naive CD20+ B"</span>: [<span class="st">"MS4A1"</span>, <span class="st">"IL4R"</span>, <span class="st">"IGHD"</span>, <span class="st">"FCRL1"</span>, <span class="st">"IGHM"</span>],</span>
<span id="cb30-12"><a href="#cb30-12" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Note IGHD and IGHM are negative markers</span></span>
<span id="cb30-13"><a href="#cb30-13" aria-hidden="true" tabindex="-1"></a>    <span class="st">"B cells"</span>: [<span class="st">"MS4A1"</span>, <span class="st">"ITGB1"</span>, <span class="st">"COL4A4"</span>, <span class="st">"PRDM1"</span>, <span class="st">"IRF4"</span>, <span class="st">"PAX5"</span>, <span class="st">"BCL11A"</span>, <span class="st">"BLK"</span>, <span class="st">"IGHD"</span>, <span class="st">"IGHM"</span>],</span>
<span id="cb30-14"><a href="#cb30-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Plasma cells"</span>: [<span class="st">"MZB1"</span>, <span class="st">"HSP90B1"</span>, <span class="st">"FNDC3B"</span>, <span class="st">"PRDM1"</span>, <span class="st">"IGKC"</span>, <span class="st">"JCHAIN"</span>],</span>
<span id="cb30-15"><a href="#cb30-15" aria-hidden="true" tabindex="-1"></a>    <span class="st">"Plasmablast"</span>: [<span class="st">"XBP1"</span>, <span class="st">"PRDM1"</span>, <span class="st">"PAX5"</span>],  <span class="co"># Note PAX5 is a negative marker</span></span>
<span id="cb30-16"><a href="#cb30-16" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD4+ T"</span>: [<span class="st">"CD4"</span>, <span class="st">"IL7R"</span>, <span class="st">"TRBC2"</span>],</span>
<span id="cb30-17"><a href="#cb30-17" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CD8+ T"</span>: [<span class="st">"CD8A"</span>, <span class="st">"CD8B"</span>, <span class="st">"GZMK"</span>, <span class="st">"GZMA"</span>, <span class="st">"CCL5"</span>, <span class="st">"GZMB"</span>, <span class="st">"GZMH"</span>, <span class="st">"GZMA"</span>],</span>
<span id="cb30-18"><a href="#cb30-18" aria-hidden="true" tabindex="-1"></a>    <span class="st">"T naive"</span>: [<span class="st">"LEF1"</span>, <span class="st">"CCR7"</span>, <span class="st">"TCF7"</span>],</span>
<span id="cb30-19"><a href="#cb30-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"pDC"</span>: [<span class="st">"GZMB"</span>, <span class="st">"IL3RA"</span>, <span class="st">"COBLL1"</span>, <span class="st">"TCF4"</span>],</span>
<span id="cb30-20"><a href="#cb30-20" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="d159a15a-9b82-496a-bedb-9e02bd5ef2b0" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> group_max(adata: sc.AnnData, groupby: <span class="bu">str</span>) <span class="op">-&gt;</span> <span class="bu">str</span>:</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>    <span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>    agg <span class="op">=</span> sc.get.aggregate(adata, by<span class="op">=</span>groupby, func<span class="op">=</span><span class="st">"mean"</span>)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> pd.Series(agg.layers[<span class="st">"mean"</span>].<span class="bu">sum</span>(<span class="dv">1</span>), agg.obs[groupby]).idxmax()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="96995645-6024-4663-ad45-79e7e3aaea1a" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>sc.pl.dotplot(adata, marker_genes, groupby<span class="op">=</span><span class="st">"leiden_res0_02"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-31-output-1.png" width="1108" height="201" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Here, we can see that cluster {eval}<code>group_max(adata[:, marker_genes["NK"]], "leiden_res0_02")</code> perhaps contains an admixture of monocytes and dendritic cells, while in cluster {eval}<code>group_max(adata[:, marker_genes["B cells"]], "leiden_res0_02")</code> we have different populations of B lymphocytes. Thus, we should perhaps consider a higher clustering resolution.</p>
<div id="a931b666-2608-463c-b8f2-a6b0d1e17209" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>sc.pl.dotplot(adata, marker_genes, groupby<span class="op">=</span><span class="st">"leiden_res0_5"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-32-output-1.png" width="1113" height="334" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>This seems like a resolution that suitable to distinguish most of the different cell types in our data. Ideally, one would look specifically into each cluster, and attempt to subcluster those if required.</p>
</section>
<section id="automatic-label-prediction" class="level3">
<h3 class="anchored" data-anchor-id="automatic-label-prediction">Automatic label prediction</h3>
<p>In addition to using marker collections to annotate our labels, there exist approaches to automatically annotate scRNA-seq datasets. One such tool is <a href="https://github.com/Teichlab/celltypist">CellTypist</a>, which uses gradient-descent optimised logistic regression classifiers to predict cell type annotations.</p>
<p>First, we need to retrive the CellTypist models that we wish to use, in this case we will use models with immune cell type and subtype populations generated using 20 tissues from 18 studies (<a href="https://www.science.org/doi/full/10.1126/science.abl5197">Domínguez Conde, et al.&nbsp;2022</a>).</p>
<div id="dc9bd354-c131-496d-b4ed-ba087ff36c60" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>ct.models.download_models(model<span class="op">=</span>[<span class="st">"Immune_All_Low.pkl"</span>], force_update<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>📜 Retrieving model list from server https://celltypist.cog.sanger.ac.uk/models/models.json
📚 Total models in list: 50
📂 Storing models in /home/runner/.celltypist/data/models
💾 Total models to download: 1
💾 Downloading model [1/1]: Immune_All_Low.pkl</code></pre>
</div>
</div>
<p>Then we predict the major cell type annotations. In this case we will enable <code>majority_voting</code>, which will assign a label to the clusters that we obtained previously.</p>
<div id="9b605f5e-bad0-4c6f-a4d0-e92307148f64" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> ct.models.Model.load(model<span class="op">=</span><span class="st">"Immune_All_Low.pkl"</span>)</span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>predictions <span class="op">=</span> ct.annotate(adata, model<span class="op">=</span><span class="st">"Immune_All_Low.pkl"</span>, majority_voting<span class="op">=</span><span class="va">True</span>, over_clustering<span class="op">=</span><span class="st">"leiden_res0_5"</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a><span class="co"># convert back to anndata||</span></span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>adata <span class="op">=</span> predictions.to_adata()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>⚠️ Warning: invalid expression matrix, expect ALL genes and log1p normalized expression to 10000 counts per cell. The prediction result may not be accurate
🔬 Input data has 16828 cells and 23427 genes
🔗 Matching reference genes in the model
🧬 5852 features used for prediction
⚖️ Scaling input data
🖋️ Predicting labels
✅ Prediction done!
🗳️ Majority voting the predictions
✅ Majority voting done!</code></pre>
</div>
</div>
<p>Let’s examine the results of automatic clustering:</p>
<div id="f5c7180a-953a-4054-a6ec-19cc0717d30d" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span><span class="st">"majority_voting"</span>, ncols<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-35-output-1.png" width="310" height="185" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Note that our previously ‘Unknown’ cluster is now assigned as ‘Pro-B cells’.</p>
</section>
<section id="annotation-with-enrichment-analysis" class="level3">
<h3 class="anchored" data-anchor-id="annotation-with-enrichment-analysis">Annotation with enrichment analysis</h3>
<p>Automatic cell type labelling with methods that require pre-trained models will not always work as smoothly, as such classifiers need to be trained on and be representitive for a given tissue and the cell types within it. So, as a more generalizable approach to annotate the cells, we can also use the marker genes from any database, for example <a href="https://panglaodb.se/">PanglaoDB</a>. Here we will use it with simple multi-variate linear regression, implemented in <a href="https://github.com/saezlab/decoupler-py">decoupler</a>. Essentially, this will test if any collection of genes are enriched in any of the cells. Ultimately, this approach is similar to many other marker-based classifiers.</p>
<p>Let’s get canonical cell markers using with <a href="https://github.com/saezlab/decoupler-py">decoupler</a> which queries the OmniPath metadata-base to obtain the <a href="https://panglaodb.se/">PanglaoDB</a> marker gene database with cannonical cell type markers.</p>
<div id="f9d52178-229a-43fe-95cb-082d8fd4c436" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Query Omnipath and get PanglaoDB</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> dc.get_resource(name<span class="op">=</span><span class="st">"PanglaoDB"</span>, organism<span class="op">=</span><span class="st">"human"</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep canonical cell type markers alone</span></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> markers[markers[<span class="st">"canonical_marker"</span>]]</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove duplicated entries</span></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>markers <span class="op">=</span> markers[<span class="op">~</span>markers.duplicated([<span class="st">"cell_type"</span>, <span class="st">"genesymbol"</span>])]</span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>markers.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Downloading data from `https://omnipathdb.org/queries/enzsub?format=json`
Downloading data from `https://omnipathdb.org/queries/interactions?format=json`
Downloading data from `https://omnipathdb.org/queries/complexes?format=json`
Downloading data from `https://omnipathdb.org/queries/annotations?format=json`
Downloading data from `https://omnipathdb.org/queries/intercell?format=json`
Downloading data from `https://omnipathdb.org/about?format=text`
Downloading data from `https://omnipathdb.org/resources?format=json`
0.00B [00:00, ?B/s]56.5kB [00:00, 168MB/s]
Downloading annotations for all proteins from the following resources: `['PanglaoDB']`
Downloading data from `https://omnipathdb.org/annotations?entity_types=protein&amp;format=tsv&amp;resources=PanglaoDB`
0.00B [00:00, ?B/s]449kB [00:00, 4.59MB/s]1.78MB [00:00, 6.86MB/s]4.76MB [00:00, 15.6MB/s]5.91MB [00:00, 15.9MB/s]</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="35">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">genesymbol</th>
<th data-quarto-table-cell-role="th">canonical_marker</th>
<th data-quarto-table-cell-role="th">cell_type</th>
<th data-quarto-table-cell-role="th">germ_layer</th>
<th data-quarto-table-cell-role="th">human</th>
<th data-quarto-table-cell-role="th">human_sensitivity</th>
<th data-quarto-table-cell-role="th">human_specificity</th>
<th data-quarto-table-cell-role="th">mouse</th>
<th data-quarto-table-cell-role="th">mouse_sensitivity</th>
<th data-quarto-table-cell-role="th">mouse_specificity</th>
<th data-quarto-table-cell-role="th">ncbi_tax_id</th>
<th data-quarto-table-cell-role="th">organ</th>
<th data-quarto-table-cell-role="th">ubiquitiousness</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>CTRB1</td>
<td>True</td>
<td>Acinar cells</td>
<td>Endoderm</td>
<td>True</td>
<td>1.000000</td>
<td>0.000629</td>
<td>True</td>
<td>0.957143</td>
<td>0.015920</td>
<td>9606</td>
<td>Pancreas</td>
<td>0.017</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>KLK1</td>
<td>True</td>
<td>Endothelial cells</td>
<td>Mesoderm</td>
<td>True</td>
<td>0.000000</td>
<td>0.008420</td>
<td>True</td>
<td>0.000000</td>
<td>0.014915</td>
<td>9606</td>
<td>Vasculature</td>
<td>0.013</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>KLK1</td>
<td>True</td>
<td>Principal cells</td>
<td>Mesoderm</td>
<td>True</td>
<td>0.000000</td>
<td>0.008145</td>
<td>True</td>
<td>0.285714</td>
<td>0.014058</td>
<td>9606</td>
<td>Kidney</td>
<td>0.013</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6</td>
<td>KLK1</td>
<td>True</td>
<td>Acinar cells</td>
<td>Endoderm</td>
<td>True</td>
<td>0.833333</td>
<td>0.005031</td>
<td>True</td>
<td>0.314286</td>
<td>0.012826</td>
<td>9606</td>
<td>Pancreas</td>
<td>0.013</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>KLK1</td>
<td>True</td>
<td>Plasmacytoid dendritic cells</td>
<td>Mesoderm</td>
<td>True</td>
<td>0.000000</td>
<td>0.008202</td>
<td>True</td>
<td>1.000000</td>
<td>0.012914</td>
<td>9606</td>
<td>Immune system</td>
<td>0.013</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="28dfae5d-010f-4ebc-b67a-8a6a436cfffa" class="cell" data-execution_count="36">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>dc.run_mlm(mat<span class="op">=</span>adata, net<span class="op">=</span>markers, weight<span class="op">=</span><span class="va">None</span>, source<span class="op">=</span><span class="st">"cell_type"</span>, target<span class="op">=</span><span class="st">"genesymbol"</span>, verbose<span class="op">=</span><span class="va">True</span>, use_raw<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Running mlm on mat with 16828 samples and 23427 targets for 118 sources.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>  0%|          | 0/2 [00:00&lt;?, ?it/s] 50%|█████     | 1/2 [00:05&lt;00:05,  5.44s/it]100%|██████████| 2/2 [00:09&lt;00:00,  4.45s/it]100%|██████████| 2/2 [00:09&lt;00:00,  4.60s/it]</code></pre>
</div>
</div>
<p>The obtained results are stored in the .obsm key, with <code>mlm_estimate</code> representing coefficient t-values:</p>
<div id="5a37b839-e747-4efc-8e57-ef06fe9ce83c" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>adata.obsm[<span class="st">"mlm_estimate"</span>].head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="37">
<div>


<table class="dataframe caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Endothelial cells</th>
<th data-quarto-table-cell-role="th">Principal cells</th>
<th data-quarto-table-cell-role="th">Acinar cells</th>
<th data-quarto-table-cell-role="th">Plasmacytoid dendritic cells</th>
<th data-quarto-table-cell-role="th">Paneth cells</th>
<th data-quarto-table-cell-role="th">Fibroblasts</th>
<th data-quarto-table-cell-role="th">Ductal cells</th>
<th data-quarto-table-cell-role="th">Macrophages</th>
<th data-quarto-table-cell-role="th">Luminal epithelial cells</th>
<th data-quarto-table-cell-role="th">Pulmonary alveolar type II cells</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">Enterochromaffin cells</th>
<th data-quarto-table-cell-role="th">Urothelial cells</th>
<th data-quarto-table-cell-role="th">Peritubular myoid cells</th>
<th data-quarto-table-cell-role="th">Germ cells</th>
<th data-quarto-table-cell-role="th">Distal tubule cells</th>
<th data-quarto-table-cell-role="th">Follicular cells</th>
<th data-quarto-table-cell-role="th">Taste receptor cells</th>
<th data-quarto-table-cell-role="th">GABAergic neurons</th>
<th data-quarto-table-cell-role="th">Meningeal cells</th>
<th data-quarto-table-cell-role="th">Merkel cells</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">AAACCCAAGGATGGCT-1</td>
<td>-0.707396</td>
<td>-0.849383</td>
<td>-0.845672</td>
<td>-0.492002</td>
<td>-1.367829</td>
<td>0.578436</td>
<td>0.238030</td>
<td>-1.056763</td>
<td>2.040976</td>
<td>-0.677132</td>
<td>...</td>
<td>-0.254653</td>
<td>-0.732992</td>
<td>-0.524543</td>
<td>-0.425869</td>
<td>0.134747</td>
<td>-0.597432</td>
<td>-0.938093</td>
<td>-0.469835</td>
<td>0.039471</td>
<td>-0.104978</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">AAACCCAAGGCCTAGA-1</td>
<td>-0.222257</td>
<td>0.336049</td>
<td>0.262638</td>
<td>0.574957</td>
<td>1.222508</td>
<td>2.367615</td>
<td>0.674121</td>
<td>-1.631422</td>
<td>1.335576</td>
<td>-0.514058</td>
<td>...</td>
<td>-0.363779</td>
<td>-1.063999</td>
<td>0.262833</td>
<td>-0.071403</td>
<td>-1.137959</td>
<td>-0.811549</td>
<td>-0.769173</td>
<td>-0.573635</td>
<td>-0.149589</td>
<td>-1.024753</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">AAACCCAAGTGAGTGC-1</td>
<td>-0.619661</td>
<td>0.646598</td>
<td>0.380288</td>
<td>-1.134570</td>
<td>0.735214</td>
<td>0.339773</td>
<td>-1.128929</td>
<td>1.977640</td>
<td>2.373118</td>
<td>-0.229449</td>
<td>...</td>
<td>-0.106915</td>
<td>-0.388254</td>
<td>-0.557494</td>
<td>1.438753</td>
<td>-0.358688</td>
<td>-0.401626</td>
<td>-0.733657</td>
<td>-0.284470</td>
<td>0.014862</td>
<td>-0.348243</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">AAACCCACAAGAGGCT-1</td>
<td>-0.259110</td>
<td>-1.807428</td>
<td>-1.449395</td>
<td>12.765751</td>
<td>-1.157377</td>
<td>2.245884</td>
<td>0.020253</td>
<td>1.010314</td>
<td>2.840580</td>
<td>-1.037617</td>
<td>...</td>
<td>-0.700302</td>
<td>-1.006705</td>
<td>2.407655</td>
<td>-0.961870</td>
<td>-0.362841</td>
<td>-0.810489</td>
<td>-0.560645</td>
<td>-0.565012</td>
<td>-0.357717</td>
<td>0.842545</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">AAACCCACATCGTGGC-1</td>
<td>0.826456</td>
<td>-0.495052</td>
<td>1.222837</td>
<td>-1.506009</td>
<td>-0.619046</td>
<td>1.809103</td>
<td>-0.611490</td>
<td>1.060724</td>
<td>-0.686973</td>
<td>0.737467</td>
<td>...</td>
<td>-0.995505</td>
<td>-0.386358</td>
<td>-0.702472</td>
<td>-0.736160</td>
<td>-0.276282</td>
<td>-0.195236</td>
<td>1.622501</td>
<td>-0.207567</td>
<td>-0.410453</td>
<td>3.025584</td>
</tr>
</tbody>
</table>

<p>5 rows × 118 columns</p>
</div>
</div>
</div>
<p>To visualize the obtianed scores, we can re-use any of scanpy’s plotting functions. First though, we will extract them from the adata object.</p>
<div id="d3bccb82-0a48-41ca-b928-b9b74d1f7ed3" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>acts <span class="op">=</span> dc.get_acts(adata<span class="op">=</span>adata, obsm_key<span class="op">=</span><span class="st">"mlm_estimate"</span>)</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>    acts,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>    color<span class="op">=</span>[</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"majority_voting"</span>,</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"B cells"</span>,</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>        <span class="st">"T cells"</span>,</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Monocytes"</span>,</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Erythroid-like and erythroid precursor cells"</span>,</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"NK cells"</span>,</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>    wspace<span class="op">=</span><span class="fl">0.5</span>,</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a>    ncols<span class="op">=</span><span class="dv">3</span>,</span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-39-output-1.png" width="749" height="361" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>These results further confirm the our automatic annotation with Celltypist. In addition, here we can also transfer the max over-representation score estimates to assign a label to each cluster.</p>
<div id="b2032395-a3ae-4705-9d94-a4f4dd10c1c3" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>mean_enr <span class="op">=</span> dc.summarize_acts(acts, groupby<span class="op">=</span><span class="st">"leiden_res0_5"</span>, min_std<span class="op">=</span><span class="fl">1.0</span>)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>annotation_dict <span class="op">=</span> dc.assign_groups(mean_enr)</span>
<span id="cb46-3"><a href="#cb46-3" aria-hidden="true" tabindex="-1"></a>adata.obs[<span class="st">"dc_anno"</span>] <span class="op">=</span> [annotation_dict[clust] <span class="cf">for</span> clust <span class="kw">in</span> adata.obs[<span class="st">"leiden_res0_5"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s compare all resulting annotations here</p>
<div id="0949dc61-20db-4343-b1d1-dc71f7c02a2b" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="st">"majority_voting"</span>, <span class="st">"dc_anno"</span>], ncols<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>... storing 'dc_anno' as categorical</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-41-output-2.png" width="340" height="361" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Great. We can see that the different approaches to annotate the data are largely concordant. Though, these annotations are decent, cell type annotation is laborous and repetitive task, one which typically requires multiple rounds of sublucstering and re-annotation. Nevertheless, we now have a good basis with which we can further proceed with manually refining our annotations.</p>
</section>
<section id="differentially-expressed-genes-as-markers" class="level3">
<h3 class="anchored" data-anchor-id="differentially-expressed-genes-as-markers">Differentially-expressed Genes as Markers</h3>
<p>Furthermore, one can also calculate marker genes per cluster and then look up whether we can link those marker genes to any known biology, such as cell types and/or states. This is typically done using simple statistical tests, such as Wilcoxon and t-test, for each cluster vs the rest.</p>
<div id="6b69a65c-199f-4883-9037-1df388b06861" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Obtain cluster-specific differentially expressed genes</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>sc.tl.rank_genes_groups(adata, groupby<span class="op">=</span><span class="st">"leiden_res0_5"</span>)</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter those</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>sc.tl.filter_rank_genes_groups(adata, min_fold_change<span class="op">=</span><span class="fl">1.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can then visualize the top 5 differentially-expressed genes on a dotplot.</p>
<div id="adf45dad-5f3d-465b-acd7-aa4f8bc42f65" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>sc.pl.rank_genes_groups_dotplot(adata, groupby<span class="op">=</span><span class="st">"leiden_res0_5"</span>, standard_scale<span class="op">=</span><span class="st">"var"</span>, n_genes<span class="op">=</span><span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>WARNING: dendrogram data not found (using key=dendrogram_leiden_res0_5). Running `sc.tl.dendrogram` with default parameters. For fine tuning it is recommended to run `sc.tl.dendrogram` independently.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-43-output-2.png" width="1203" height="301" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We see that <em>LYZ</em>, <em>ACT8</em>, <em>S100A6</em>, <em>S100A4</em>, and <em>CST3</em> are all highly expressed in cluster <code>3</code>. Let’s visualize those at the UMAP space:</p>
<div id="d849f8b3-c085-4634-9272-028563d4629e" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>cluster3_genes <span class="op">=</span> [<span class="st">"LYZ"</span>, <span class="st">"ACTB"</span>, <span class="st">"S100A6"</span>, <span class="st">"S100A4"</span>, <span class="st">"CST3"</span>]</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>sc.pl.umap(adata, color<span class="op">=</span>[<span class="op">*</span>cluster3_genes, <span class="st">"leiden_res0_5"</span>], legend_loc<span class="op">=</span><span class="st">"on data"</span>, frameon<span class="op">=</span><span class="va">False</span>, ncols<span class="op">=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-44-output-1.png" width="589" height="355" class="figure-img"></p>
</figure>
</div>
</div>
</div>
<p>Similarly, we can also generate a Violin plot with the distrubtions of the same genes across the clusters.</p>
<div id="67544d51-1e90-4329-a800-b6c6d043a6cb" class="cell" data-scrolled="true" data-execution_count="44">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>sc.pl.violin(adata, keys<span class="op">=</span>cluster3_genes[<span class="dv">0</span>:<span class="dv">3</span>], groupby<span class="op">=</span><span class="st">"leiden_res0_5"</span>, multi_panel<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.
Using categorical units to plot a list of strings that are all parsable as floats or dates. If these strings should be plotted as numbers, cast to the appropriate data type before plotting.</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="scanpy-basic-scrna-tutorial_files/figure-html/cell-45-output-2.png" width="540" height="191" class="figure-img"></p>
</figure>
</div>
</div>
</div>


</section>
</section>

<p>© <a href="https://hakyimlab.org">HakyImLab and Listed Authors</a> - <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0 License</a></p></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>